---
title: "DecorateInsertions.Rmd"
author: "bhaas"
date: '2023-04-07'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
```


# add hotspot info

```{r}

insertions_data = read.table("../HotspotsRevisited/hotspots.win_1e5.tsv.w_20_neighbors.regrouped_by_insert_gene", header=T, sep="\t", stringsAsFactors = F) %>%
   
  select(sample,
         sample_id,
         participant,
         tumor_or_normal,
         cohort,
         project,
         seqtype,
         insertion = contig,
         virus,
         virus_coord,
         virus_brkend_grp,
         repr_virus_breakend,
         humanchr,
         human_coord,
         is_primary,
         split,
         span,
         total,
         total_rpm,
         splice_type,
         mapped,
         n_bases_covered,
         mapped_rpm,
         hotspot = gene_regrouped_hotspot_name, 
         hotspot_participant_count = gene_regrouped_hotspot_size)


nrow(insertions_data)
nrow(insertions_data %>% unique())

```


# add human fusions from STAR-Fusion

```{r}

fusions_at_insertions = read.table("../Insertion_STARF_Fusion_Mapping/FI_fusions_at_insertions.tsv", header=T, sep="\t", stringsAsFactors = F)

fusions_at_insertions = fusions_at_insertions %>% select(sample_id, contig, fusion_name) %>% unique() %>%
  group_by(sample_id, contig) %>% arrange(fusion_name) %>%
  mutate(fusions_at_insertion = paste(fusion_name, collapse=";")) %>%
  select(sample_id, contig, fusions_at_insertion) %>% unique()

fusions_at_insertions
nrow(fusions_at_insertions)
fusions_at_insertions %>% filter(grepl(";", fusions_at_insertion))
```



```{r}
insertions_data = left_join(insertions_data, 
                            fusions_at_insertions,
                            by=c('sample_id'='sample_id', 
                                 'insertion' = 'contig'))

nrow(insertions_data)
nrow(insertions_data %>% filter(! is.na(fusions_at_insertion)))
```


# add gene-spliced virus/human chimeric transcripts

```{r}

spliced_insertions_info = read.table("../Insertion_Spliced_Human/spliced_insertions.tsv", header=T, sep="\t", stringsAsFactors = F) %>%
  select(sample_id, contig, seqtype, spliced_gene=Gene_ID)

nrow(spliced_insertions_info)

insertions_data = left_join(insertions_data,
                            spliced_insertions_info,
                            by=c('sample_id', 'insertion'='contig', 'seqtype'))
nrow(insertions_data)

nrow(insertions_data %>% filter(! is.na(spliced_gene)) )

```


# add CNV info

```{r}

cnv_info = read.table("../Insertion_and_CNVs/TCGA_insertions_and_CNV_within10kb", header=T, sep="\t", stringsAsFactors = F) %>% unique()


cnv_info = cnv_info %>% select(sample_id = sample, insertion, cnv_lend, cnv_rend, copy_number)
nrow(cnv_info)

insertions_data = left_join(insertions_data,
                            cnv_info,
                            by=c('sample_id', 'insertion'))

nrow(insertions_data)
nrow(cnv_info %>% filter(! is.na(copy_number)))
```


# add Expression data

```{r}

expr_info = read.table("../Insertion_and_EXPR_effects/virus_insertions_unfiltered.10kb.expression_region_analysis.tsv", header=T, sep="\t", stringsAsFactors = F)

expr_info = expr_info %>% select(sample_id, insertion=contig, seqtype, expr_quantile) %>% unique()
nrow(expr_info)

nrow(insertions_data)
insertions_data = left_join(insertions_data,
                            expr_info,
                            by=c('sample_id', 'insertion', 'seqtype'))
nrow(insertions_data)

nrow(insertions_data %>% filter(! is.na(expr_quantile)) )
```



# basic auditing:

```{r}
nrow(insertions_data)
nrow(insertions_data %>% filter(! is.na(fusions_at_insertion)))
nrow(insertions_data %>% filter(! is.na(spliced_gene)))
nrow(insertions_data %>% filter(! is.na(copy_number)))
nrow(insertions_data %>% filter(! is.na(expr_quantile)))


```




```{r}

write.table(insertions_data, file="all_insertions_and_virus_content_merged.FILTERED.Decorated.tsv", sep="\t", row.names=F, quote=F)


```

