---
title: "CNV_and_expr_at_insertions"
author: "bhaas"
date: '2022-11-04'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
```


# examine CNV
```{r}

insertion_cnv_mappings = read.table("filtered_insertions.nr.tsv.CNV_within_10kb", header=T, sep="\t", stringsAsFactors = F)


```


```{r}
# remove the likely contaminants
nrow(insertion_cnv_mappings)

contaminants_info = read.table("../Find_Likely_Contaminants/TCGA_VIF_contaminants_only.tsv.saved", header=T, sep="\t", stringsAsFactors = F)

insertion_cnv_mappings = left_join(insertion_cnv_mappings, contaminants_info, by=c('sample'='sample_name', 'virus')) %>%  filter(is.na(CONTAM)) %>% select(-CONTAM)
  
nrow(insertion_cnv_mappings)

```


```{r}

insertion_cnv_mappings %>% group_by(TCGA, virus) %>% arrange(desc(copy_number)) %>% mutate(cnv_ranking = row_number()) %>%
  ungroup() %>%
  ggplot(aes(x=cnv_ranking, y=copy_number, color=virus)) + geom_point() + geom_line() +
  facet_wrap(~TCGA, scale='free')


```

```{r}
# restrict plot to just CESC, LIHC, and HNSC

insertion_cnv_mappings %>% filter(TCGA %in% c('CESC', 'HNSC', 'LIHC') ) %>% group_by(TCGA, virus) %>% arrange(desc(copy_number)) %>% mutate(cnv_ranking = row_number()) %>%
  ungroup() %>%
  mutate(tcga_virus = paste0(TCGA, "-", virus)) %>%
  ggplot(aes(x=cnv_ranking, y=copy_number, color=virus)) + geom_point() + geom_line() +
  facet_wrap(~tcga_virus, scale='free_x') + 
  geom_hline(yintercept=2, linetype='dotted', color='black')


insertion_cnv_mappings %>% filter(TCGA %in% c('CESC', 'HNSC', 'LIHC') ) %>% group_by(TCGA, virus) %>% arrange(desc(copy_number)) %>% mutate(cnv_ranking = row_number()) %>%
  ungroup() %>%
  mutate(tcga_virus = paste0(TCGA, "-", virus)) %>%
  ggplot(aes(x=cnv_ranking, y=copy_number, color=virus)) + geom_point() + geom_line() +
  facet_wrap(~TCGA, scale='free_x') + 
  geom_hline(yintercept=2, linetype='dotted', color='black')



```





```{r}


insertion_cnv_mappings %>% mutate(tcga_virus = paste0(TCGA, "-", virus)) %>%
 ggplot(aes(x=tcga_virus, y=copy_number, color=virus)) + 
  geom_boxplot() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  geom_hline(yintercept = 2, color='black', linetype='dotted')
  


```

```{r}

# just focus on the most relevant ones.

insertion_cnv_mappings %>% mutate(tcga_virus = paste0(TCGA, "-", virus)) %>%
  filter(tcga_virus %in% c('CESC-HPV16', 'HNSC-HPV16', 'LIHC-NC_003977_3182nt_Hepatitis_B_virus')) %>%
 ggplot(aes(x=tcga_virus, y=copy_number, color=virus)) + 
  geom_boxplot() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  geom_hline(yintercept = 2, color='black', linetype='dotted')
  



```




# examine expression rankings for neighboring genes.

```{r}

insertion_expr_info = read.table("filtered_insertions.nr.tsv.w_10neighbors.expr_vals", header=T, sep="\t", stringsAsFactors = F)

insertion_expr_info_min_rank = insertion_expr_info %>% group_by(sample, insertion) %>% arrange(rank) %>% filter(row_number() == 1) %>% ungroup()

```


```{r}

insertion_expr_info_min_rank %>% mutate(tcga_virus = paste0(TCGA, "-", virus)) %>% 
  ggplot(aes(x = tcga_virus, y=-1*rank, color=virus)) + geom_boxplot() +
   theme(axis.text.x = element_text(angle = 90, hjust = 1)) 

```


```{r}

insertion_expr_info_min_rank %>% mutate(tcga_virus = paste0(TCGA, "-", virus)) %>% 
   filter(tcga_virus %in% c('CESC-HPV16', 'HNSC-HPV16', 'LIHC-NC_003977_3182nt_Hepatitis_B_virus')) %>%
  ggplot(aes(x = tcga_virus, y=-1*rank, color=virus)) + geom_boxplot() +
   theme(axis.text.x = element_text(angle = 90, hjust = 1)) 




```





# Do Expr rankings correlate with CNV levels?


```{r}

expr_and_cnv_info = full_join(insertion_expr_info_min_rank %>% select(TCGA, sample, virus, insertion, rank),
                              insertion_cnv_mappings %>% select(TCGA, sample, virus, insertion, copy_number),
                              by=c('TCGA', 'sample', 'virus', 'insertion')
                              )



```


```{r}
expr_and_cnv_info %>% ggplot(aes(x=copy_number, y=rank)) + geom_point()


```


```{r}


smoothScatter(expr_and_cnv_info$copy_number, -1*expr_and_cnv_info$rank, xlim=c(0,20), ylim=c(0,-20))

```

```{r}

cor(expr_and_cnv_info$copy_number, -1*expr_and_cnv_info$rank, use='complete', method='spearman')
```


```{r}

cor.test(expr_and_cnv_info$copy_number, -1*expr_and_cnv_info$rank, use='complete', method='spearman')

```

```{r}

# restrict to CESC-HPV16

CESC_HPV16_cnv_expr_info = expr_and_cnv_info %>% filter(TCGA=="CESC" & virus == "HPV16") 

cor.test(CESC_HPV16_cnv_expr_info$copy_number, -1*CESC_HPV16_cnv_expr_info$rank, use='complete', method='spearman')


```

```{r}
# restrict to CESC-HPV16

CESC_HPV18_cnv_expr_info = expr_and_cnv_info %>% filter(TCGA=="CESC" & virus == "HPV18") 
cor.test(CESC_HPV18_cnv_expr_info$copy_number, -1*CESC_HPV18_cnv_expr_info$rank, use='complete', method='spearman')


```


```{r}

# restrict to LIHC-HBV

LIHC_HBV_cnv_expr_info = expr_and_cnv_info %>% filter(TCGA=="LIHC" & virus == "NC_003977_3182nt_Hepatitis_B_virus") 
cor.test(LIHC_HBV_cnv_expr_info$copy_number, -1*LIHC_HBV_cnv_expr_info$rank, use='complete', method='spearman')


```



# examine FN1 in liver

```{r}


FN1_hotspot_insertions = read.table("../../InsertionHotspots/hotspots.w_20neighbors.tsv", header=T, sep="\t", stringsAsFactors = F) %>% filter(hotspot== "chr2:215425286") %>% select(sample, contig)

FN1_hotspot_insertions 


```

```{r}
left_join(FN1_hotspot_insertions,
          expr_and_cnv_info,
          by=c('sample', 'contig'='insertion') ) %>% filter(! is.na(virus))
          
```


# Examine counts of insertions per sample.

```{r}


nr_insertions = read.table("filtered_insertions.nr.tsv", header=T, sep="\t", stringsAsFactors = F)

nr_insertions$TCGA = sapply(nr_insertions$sample, function(x) { str_split(x, "-")[[1]][1]; })


```


```{r}

sample_insertion_counts = nr_insertions %>% group_by(TCGA, sample, virus) %>% tally(name='sample_insertion_count') %>% ungroup()

head(sample_insertion_counts)
```

```{r}

sample_insertion_counts %>% group_by(TCGA, virus) %>% arrange(desc(sample_insertion_count)) %>% mutate(ordered_sample = row_number()) %>% ungroup() %>%
  ggplot(aes(x=ordered_sample, y=sample_insertion_count, color=TCGA)) +
  geom_point() +
  facet_wrap(~virus, scale='free')

```


```{r}
# restrict above to just the main TCGA insertion types: CESC, HNSC, LIHC

sample_insertion_counts %>% 
  filter(TCGA %in% c("CESC", "LIHC", "HNSC") ) %>%
  group_by(TCGA, virus) %>% arrange(desc(sample_insertion_count)) %>% mutate(ordered_sample = row_number()) %>% ungroup() %>%
  ggplot(aes(x=ordered_sample, y=sample_insertion_count, color=virus)) +
  geom_point() + geom_line() +
  facet_wrap(~TCGA, scale='free')


```

```{r}

CESC_HPV16_sample_insertion_counts  =  sample_insertion_counts %>% filter(TCGA == "CESC" & virus == "HPV16") %>% pull(sample_insertion_count)

CESC_HPV18_sample_insertion_counts  =  sample_insertion_counts %>% filter(TCGA == "CESC" & virus == "HPV18") %>% pull(sample_insertion_count)

HNSC_HPV16_sample_insertion_counts = sample_insertion_counts %>% filter(TCGA == "HNSC" & virus == "HPV16") %>% pull(sample_insertion_count)

LIHC_HBV_sample_insertion_counts = sample_insertion_counts %>% filter(TCGA == "LIHC" & virus == "NC_003977_3182nt_Hepatitis_B_virus") %>% pull(sample_insertion_count)



```

```{r}

wilcox.test(CESC_HPV16_sample_insertion_counts, LIHC_HBV_sample_insertion_counts)

wilcox.test(CESC_HPV18_sample_insertion_counts, LIHC_HBV_sample_insertion_counts)

wilcox.test(CESC_HPV16_sample_insertion_counts, CESC_HPV18_sample_insertion_counts)

wilcox.test(CESC_HPV16_sample_insertion_counts, HNSC_HPV16_sample_insertion_counts)

```






```{r}


sample_insertion_counts %>% 
  filter(TCGA %in% c("CESC", "LIHC", "HNSC") ) %>%
   mutate(tcga_virus = paste0(TCGA, "-", virus)) %>%
   ggplot() + geom_boxplot(aes(x=tcga_virus, y=sample_insertion_count, color=virus)) +
   theme(axis.text.x = element_text(angle = 90, hjust = 1)) 
  


```


```{r}

sample_insertion_counts %>% 
  filter(TCGA %in% c("CESC", "LIHC", "HNSC") ) %>%
   mutate(tcga_virus = paste0(TCGA, "-", virus)) %>%
  group_by(tcga_virus) %>%
  summarize(median_sample_insertion_count = median(sample_insertion_count)) %>%
  arrange(desc(median_sample_insertion_count))


```



# Examine insertion read support per sample (max where mult insertions)

```{r}

top_each_sample_nr_insertions = nr_insertions  %>% group_by(sample, virus) %>% arrange(desc(total)) %>% filter(row_number() == 1) %>% ungroup() %>%
  mutate(tcga_virus = paste0(TCGA, "-", virus)) 

top_each_sample_nr_insertions %>%
  ggplot(aes(x=tcga_virus, y=total, color=virus)) + geom_boxplot() +
   theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  ggtitle("Dist of reads supporting top-most virus insertion per sample")
 



```


```{r}


top_each_sample_nr_insertions %>%
  filter(TCGA %in% c('CESC', 'LIHC', 'HNSC')) %>%
  ggplot(aes(x=tcga_virus, y=total, color=virus)) + geom_boxplot() +
   theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  ggtitle("Dist of reads supporting top-most virus insertion per sample") +
  scale_y_log10()

 
```



```{r}
# CESC quantiles
quantile(top_each_sample_nr_insertions %>% filter(TCGA == "CESC") %>% pull('total'), probs = seq(0,1,.01))

```



