---
title: "VirusExpressionAlongGenome"
author: "Brian Haas"
date: '2022-06-04'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
```

```{r}

alldata = NULL

files = list.files("data/LIHC/", pattern = "bedgraph")
for (file in files) {
    message("processing file: ", file)
    path = paste0("data/LIHC/", file)
    samplename = str_replace(file, ".VirusDetect.igvjs.bam.bedgraph", "")
    data = read.table(path, sep="\t", stringsAsFactors = F, header=F)
    colnames(data) = c('chrom', 'lend', 'rend', 'coverage')
    data = data %>% filter(grepl("Hepatitis_B_virus", chrom))
    # restrict to those HPVs with min coverage
    covered_hpvs = data %>% group_by(chrom) %>% summarize(sumcov = sum(coverage)) %>% filter(sumcov >= 1000) %>% pull(chrom)
    data = data %>% filter(chrom %in% covered_hpvs)
    if (nrow(data) > 0) {
        data$samplename = samplename
        alldata = bind_rows(alldata, data)
        
    }    
}

```


```{r}

write.table(alldata, "LIHC.virus_coverage_info.tsv", sep="\t", quote=F)

```


# examine virus coverage

```{r}

viruses = alldata %>% select(chrom) %>% unique() %>% pull(chrom)


```

```{r}

insertions = read.table("../filtered_insertions.tsv", header=T, sep="\t", stringsAsFactors = F) %>%
  select(sample_name, virus) %>% filter(grepl("Hepatitis_B_virus", virus)) %>% unique() %>% mutate(has_insertion = TRUE)

insertions %>% head()

```



```{r}

pdf("LIHC_HBV_genome_coverage.pdf")

for (virus in viruses) {
  
  virus_data =  alldata %>% filter(chrom == virus)
  
  
   virus_coverage_matrix = virus_data %>% select(lend, coverage, samplename) %>% spread(key=lend, value=coverage, fill = 0)
  samplenames = virus_coverage_matrix$samplename
  rownames(virus_coverage_matrix) = samplenames
  virus_coverage_matrix = virus_coverage_matrix %>% select(-samplename)
  virus_coverage_matrix_filename = paste0("LIHC.", virus, ".coverage.matrix")
  write.table(virus_coverage_matrix, file=virus_coverage_matrix_filename, quote=F, sep="\t")
  
  
  virus_data = left_join(virus_data, 
                         insertions, 
                         by=c('samplename' = 'sample_name', 'chrom' = 'virus')) %>% 
    mutate(has_insertion = ifelse(is.na(has_insertion), FALSE, has_insertion))
  
  
    sample_insertion_labeling = virus_data %>% select(has_insertion, samplename) %>% unique() %>% arrange(has_insertion)
  sample_insertion_labeling_filename = paste0("LIHC.", virus, ".insertion_labeling.tsv")
  write.table(sample_insertion_labeling, file=sample_insertion_labeling_filename, sep="\t", quote=F, row.names = F, col.names = F)
  
  
  make_ptr_heatmap_cmd = paste0("~/GITHUB/trinityrnaseq/Analysis/DifferentialExpression/PtR -m ", 
                                virus_coverage_matrix_filename,
                                " --gene_factors ",
                                sample_insertion_labeling_filename,
                                " --heatmap --sample_clust none --log2 --heatmap_colorscheme \"black,yellow\"")
  system(make_ptr_heatmap_cmd)
 

  
  
  p = virus_data %>% ggplot(aes(x=lend, y=samplename, fill=log2(coverage+1))) + geom_tile() + ggtitle(virus) +
    facet_wrap(~has_insertion, ncol=1, scale='free_y')
  plot(p)
  
}

dev.off()

plot(p)
```


