---
title: "InsertionAnalysis.by_orthogonal_breakpoint_support.Rmd"
author: "bhaas"
date: '2023-03-07'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
knitr::opts_chunk$set(warning = FALSE, message = FALSE) 

```


```{r}

insertions_data = read.table("all_insertions_and_virus_content_merged.w_shared_site_info.tsv", header=T, sep="\t", stringsAsFactors = F)

insertions_data = insertions_data %>% mutate(participant_virus = paste(participant, virus))

nrow(insertions_data)

```

```{r}

# Examine rna-seq insertions with matched WXS or WGS

rnaseq_participants = insertions_data %>% filter(seqtype == "RNA") %>% select(participant) %>% unique() %>% pull(participant)
other_participants = insertions_data %>%  filter(seqtype != "RNA") %>% select(participant) %>% unique() %>% pull(participant)

# identify participants with both
participants_both = intersect(rnaseq_participants, other_participants)

participants_both
length(participants_both)
```

```{r}
# how many insertions are in these samples?

insertions_data %>% filter(participant %in% participants_both) %>% nrow()


```

6251 / 37184 = 17%

```{r}

# WXS paired w/ rna-seq
wxs_participants = insertions_data %>% filter(seqtype == "WXS") %>% select(participant) %>% unique() %>% pull(participant)



intersect(rnaseq_participants, wxs_participants)

length(intersect(rnaseq_participants, wxs_participants))

```


```{r}

# WGS paired w/ rna-seq
wgs_participants = insertions_data %>% filter(seqtype == "WGS") %>% select(participant) %>% unique() %>% pull(participant)



intersect(rnaseq_participants, wgs_participants)

length(intersect(rnaseq_participants, wgs_participants))


```


```{r}

# how many with all three: rna, wxs, and wgs?

intersect(intersect(rnaseq_participants, wxs_participants), wgs_participants)


```


```{r}

library(GenomicRanges)

df_A_with_nearest_df_B = function(df_A, df_B) {
  
    # each df requires fields:  Chromosome_val, Start_val, End_val
  
    gr_df_A = with(df_A, GRanges(Chromosome_val,
                                IRanges(start=Start_val, end=End_val)))
    
    gr_df_B = with(df_B, GRanges(Chromosome_val,
                                IRanges(start=Start_val, end=End_val)))

  
    nearest_df_B = df_B[nearest(gr_df_A, gr_df_B, ignore.strand=T),]
    
    df_A_with_nearest_df_B = bind_cols(df_A %>% select(-Chromosome_val, -Start_val, -End_val), 
                                       nearest_df_B %>% select(-Chromosome_val, -Start_val, -End_val))
    
    return(df_A_with_nearest_df_B)

}


```

```{r}


rnaseq_paired_data = insertions_data %>% filter(seqtype=='RNA') %>% filter(participant %in% participants_both)

other_paired_data =  insertions_data %>% filter(seqtype!='RNA') %>% filter(participant %in% participants_both)



rnaseq_paired_data = rnaseq_paired_data %>% mutate(Chromosome_val = humanchr, Start_val = human_coord, End_val = human_coord) %>%
  select(rna_sample_name = sample_name, rna_contig = contig, rna_humanchr = humanchr,
         rna_human_coord = human_coord, rna_virus_brkend_grp = virus_brkend_grp,
         rna_cohort = cohort, rna_project = project, rna_participant = participant, rna_sample_id = sample_id,
         rna_total = total, rna_total_rpm = total_rpm, Chromosome_val, Start_val, End_val, participant, virus, rna_tumor_or_normal = tumor_or_normal,
         rnaseq_participant_virus = participant_virus)

other_paired_data = other_paired_data %>% mutate(Chromosome_val = humanchr, Start_val = human_coord, End_val = human_coord) %>%
  select(other_sample_name = sample_name, other_contig = contig, 
         other_humanchr = humanchr, other_human_coord = human_coord, other_seqtype = seqtype, 
         other_cohort = cohort, other_project = project, other_participant = participant, other_sample_id = sample_id,
         other_virus_brkend_grp = virus_brkend_grp, other_total = total, other_total_rpm = total_rpm, other_tumor_or_normal = tumor_or_normal,
         Chromosome_val, Start_val, End_val, other_participant_virus = participant_virus)


```

```{r}

participants_virus_both = intersect(rnaseq_paired_data %>% pull(rnaseq_participant_virus),
                                    other_paired_data %>% pull(other_participant_virus) )



rna_nearest_data = NULL

for (participant_val in participants_virus_both) {
  
  rna_paired_data_participant = rnaseq_paired_data %>% filter(rnaseq_participant_virus == participant_val) 
  
  other_paired_data_participant = other_paired_data %>% filter(other_participant_virus == participant_val)
  
  rna_nearest_other_participant = df_A_with_nearest_df_B(rna_paired_data_participant, other_paired_data_participant)
  
  rna_nearest_data = bind_rows(rna_nearest_data, rna_nearest_other_participant)
  
}


```


```{r}

rna_nearest_data = rna_nearest_data %>% mutate(delta = abs(rna_human_coord - other_human_coord))

```

```{r}

# count number of insertions in range

dist_ranges = c(0, 1, 5, 10, 20, 50, 100, 500, 1000, 2000, 3000, 4000, 5000, 6000, 7000, 8000, 9000, 10000, 20000, 50000, 100000)

insertion_count_vs_dist_range_df = NULL

for (dist_range in dist_ranges) {
  
    num_insertions = rna_nearest_data %>% filter(delta <= dist_range) %>% select(rna_participant, rna_contig, rna_cohort, rna_project) %>% unique() %>% nrow()
  
    insertion_count_vs_dist_range_df = bind_rows(insertion_count_vs_dist_range_df, data.frame(dist_range = dist_range, insertion_count = num_insertions))
  
}

insertion_count_vs_dist_range_df

```

```{r}

insertion_count_vs_dist_range_df %>% ggplot(aes(x=dist_range, y=insertion_count)) + geom_point() +
  ggtitle("insertion count ~ allowed dist between breakpoints") + 
  geom_vline(xintercept=10000, color='red')


```



```{r}
WITHIN_RANGE_DIST = 10000

rna_nearest_data = rna_nearest_data %>% mutate(within_range = delta <= WITHIN_RANGE_DIST)

# count them

num_rna_insertions_validated = rna_nearest_data %>% filter(within_range) %>% select(rna_participant, rna_contig, rna_cohort, rna_project) %>% unique() %>% nrow()
message('num rna insertions validated within dist: ', num_rna_insertions_validated)

num_other_insertions_validated = rna_nearest_data %>% filter(within_range) %>% select(other_participant, other_contig, other_cohort, other_project, other_seqtype) %>% unique() %>% nrow()

message('num other insertions validated within dist: ' , num_other_insertions_validated)

# breakdown of others
rna_nearest_data %>% filter(within_range) %>% select(other_participant, other_contig, other_cohort, other_project, other_seqtype) %>% unique() %>% group_by(other_seqtype, other_cohort, other_project) %>% tally(name='num_other_insertions')


```

```{r}

write.table(rna_nearest_data, file="rna_insertions_within_10kb_other_seqtype_insertions.tsv", quote=F, sep="\t", row.names=F)


```


```{r}
# how many participants had orthogonal validations above?

rna_nearest_data %>% select(other_participant) %>% unique() %>% nrow()

```

234 / 252 = 93%
```{r}

# were any of the rna_nearest_data previously labeled as suspicious?

inner_join(rna_nearest_data %>% 
             select(rna_sample_id, rna_cohort, rna_project, virus, rna_humanchr, rna_human_coord, rna_virus_brkend_grp),
           
           insertions_data %>% filter(seqtype == "RNA") %>%
             filter(matching_virus_content & suspicious) %>%
             select(sample_id, cohort, project, virus, humanchr, human_coord, virus_brkend_grp, insertion_token),
           
           by=c('rna_sample_id'='sample_id',
                'rna_cohort'='cohort', 
                'rna_project'='project',
                'virus'='virus',
                'rna_humanchr'='humanchr',
                'rna_human_coord'='human_coord',
                'rna_virus_brkend_grp'='virus_brkend_grp') )


```

- only 2 and these are splice types.


```{r}

# WXS within range:

rna_with_wxs_support = rna_nearest_data %>% 
  filter(other_seqtype == "WXS") %>%
  filter(within_range) %>%
  filter(! is.na(virus)) 


rna_with_wxs_support %>%
  mutate(virus = ifelse(grepl("HPV", virus), "HPV", virus)) %>%
  ggplot(aes(y=rna_total_rpm, x=other_total_rpm, color=virus)) + geom_point() +
  facet_wrap(~virus) + ggtitle("RNA insertions with WXS support")


```


```{r}

# WGS within range:

rna_with_WGS_support = rna_nearest_data %>% 
  filter(other_seqtype == "WGS") %>%
  filter(within_range) %>%
  filter(! is.na(virus))

rna_with_WGS_support %>%
  mutate(virus = ifelse(grepl("HPV", virus), "HPV", virus)) %>%
  ggplot(aes(y=rna_total_rpm, x=other_total_rpm, color=virus)) + geom_point() +
  facet_wrap(~virus) + ggtitle("RNA insertions with WGS support")


```

```{r}

# HYB within range:

rna_with_hyb_support = rna_nearest_data %>% 
  filter(other_seqtype == "HYB") %>%
  filter(within_range) 


rna_with_hyb_support %>%
  filter(! is.na(virus)) %>%
  mutate(virus = ifelse(grepl("HPV", virus), "HPV", virus)) %>%
  ggplot(aes(y=rna_total_rpm, x=other_total_rpm, color=virus)) + geom_point() +
  facet_wrap(~virus) + ggtitle("RNA insertions with HYB support")


```

```{r}



retrieve_columns = function(input_data) {

  retrieved_data = input_data %>% select(rna_sample_name, virus, rna_virus_brkend_grp,  rna_humanchr, rna_human_coord, rna_total, rna_total_rpm,
                               other_seqtype, other_humanchr, other_human_coord, other_virus_brkend_grp, other_total, other_total_rpm, delta, other_cohort,  other_tumor_or_normal)

  return(retrieved_data)
}


rnaseq_orthogonally_matched_data = bind_rows(retrieve_columns(rna_with_wxs_support),
                                             retrieve_columns(rna_with_WGS_support),
                                             retrieve_columns(rna_with_hyb_support))

rnaseq_orthogonally_matched_data

```




```{r}
## simplify according to rna_virus_brkend_grp, choosing min delta

rnaseq_orthogonally_matched_data_distilled = rnaseq_orthogonally_matched_data %>% group_by(rna_sample_name, rna_virus_brkend_grp) %>% arrange(delta) %>% filter(row_number()==1) %>% ungroup()


rnaseq_orthogonally_matched_data_distilled

```


```{r}

rnaseq_orthogonally_matched_data_distilled %>% ggplot(aes(x=rna_total_rpm, y=other_total_rpm, color=other_seqtype, shape=other_cohort)) + geom_point() + facet_wrap(~other_seqtype) +
  ggtitle("RNA-seq insertions corroborated by orthogonal sequencing")


```




```{r}

# what's the RNA rpm ranges for orthogonally validated breakpoints?

rnaseq_orthogonally_matched_data_distilled %>%
  ggplot(aes(x=other_seqtype, y=rna_total_rpm, color=other_seqtype)) + geom_boxplot() + 
  coord_cartesian(ylim=c(0,10)) + ggtitle("RNA-seq orthogonally validated insertion total_rpm distribution")
  

```

```{r}

# quantile on read count support
quantile(rnaseq_orthogonally_matched_data_distilled %>% 
           pull(rna_total), seq(0,1,0.01))

# quantile on read rpm support
quantile(rnaseq_orthogonally_matched_data_distilled %>% 
           pull(rna_total_rpm), seq(0,1,0.01))

# 97% and 94% have at least 2 and 3 evidence reads supporting, corresponding to ~ 0.02 to 0.03 chimeric rpm

```
























