---
title: "VirusBreakpointAnalysis"
author: "bhaas"
date: '2022-11-01'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(GenomicRanges)
library(tidyverse)
```

```{r}

data = read.table("../filtered_insertions.tsv", header=T, sep="\t", stringsAsFactors = F)


```

```{r}

data = data %>% mutate(virus_coord = ifelse(virus==chrA, coordA, coordB))

data = data %>% mutate(human_coord = ifelse(virus==chrA, coordB, coordA))


```


```{r}

# restrict to primary breakpoint
data = data %>% filter(is_primary == "True")

```


```{r}

sum_virus_read_coord_stats = data %>% group_by(sample_name, virus, virus_coord, splice_type) %>% summarize(sum_reads = sum(total)) %>% ungroup()

sum_virus_read_coord_stats
```


```{r}

fractional_virus_coord_stats = left_join(sum_virus_read_coord_stats, 
                                         data %>% select(sample_name, virus, virus_coord, splice_type, total),
                                         by=c('sample_name', 'virus', 'virus_coord', 'splice_type') ) %>% 
  mutate(f_reads = total / sum_reads)

fractional_virus_coord_stats

```

```{r}

virus_breakpoint_count_stats = fractional_virus_coord_stats %>% group_by(virus) %>% tally() %>% ungroup()

virus_breakpoint_count_stats

```

```{r}

sum_frac_read_support = fractional_virus_coord_stats %>% group_by(virus, virus_coord, splice_type) %>% summarize(sum_support = sum(f_reads)) %>% ungroup()
 
sum_frac_read_support %>% arrange(desc(sum_support))
```


```{r}

virus_splicing_info = read.table("data/HPV_splice_sites.txt", header=T, sep="\t", stringsAsFactors = F)

colnames(virus_splicing_info) = str_replace_all(colnames(virus_splicing_info), "\\.+", "_")

virus_splicing_info = tibble(virus_splicing_info) %>% rename(virus=HPV_Type) %>% gather(key='splice', value='coord', -virus)


```



```{r}

min_virus_brkpts = 20
viruses_of_interest = virus_breakpoint_count_stats %>% filter(n>= min_virus_brkpts) %>% pull(virus)

breakpoint_plot = sum_frac_read_support %>% filter(virus %in% viruses_of_interest) %>%
  ggplot(aes(x=virus_coord, y=sum_support, color=splice_type)) + geom_segment(aes(xend=virus_coord, yend = 0)) +
  geom_point(data=virus_splicing_info %>% filter(virus %in% viruses_of_interest), aes(x=coord, y=0), color='purple', size=rel(0.7)) +
  facet_wrap(~virus, scale='free', ncol=1)

breakpoint_plot

```

```{r}

ggsave(breakpoint_plot, file='TCGA_viral_genome_breakpoints_plot.pdf', width=8, height=11)

```



```{r}

# examine HBV

sum_frac_read_support %>% filter(virus == "NC_003977_3182nt_Hepatitis_B_virus" ) %>%
  ggplot(aes(x=virus_coord, y=sum_support, color=splice_type)) + geom_segment(aes(xend=virus_coord, yend = 0)) +
   xlim(c(1740, 1850) )



```

```{r}

data %>% filter(virus == "NC_003977_3182nt_Hepatitis_B_virus") %>% filter(virus_coord >= 1740 & virus_coord <= 1850) %>%
  select(virus_coord, sample_name, total, splice_type, entropyA, entropyB, flankA_fU, flankB_fU) %>% arrange(virus_coord)


```

```{r}

# examine HBV 
sum_frac_read_support %>% filter(virus == "NC_003977_3182nt_Hepatitis_B_virus") %>% arrange(desc(sum_support))


```


# map all virus breakpoints in human genome to nearest exons to examine ref gene splice connections

```{r}


ref_annot_exons = read.table(gzfile("../../data/ref_annot.gtf.exon_features.gz"), header=F, sep="\t", stringsAsFactors = F)

colnames(ref_annot_exons) = c('chrom', 'source', 'exon', 'lend', 'rend', 'trash1', 'strand', 'trash2', 'gene_info')
ref_annot_exons$gene_info = str_replace(ref_annot_exons$gene_info,"gene_id ", "")
```

```{r}
gr_exons = with(ref_annot_exons, GRanges(chrom, IRanges(start=lend, end=rend)) )#, strand=strand))

gr_insertions = with(data, GRanges(seqnames=humanchr, ranges=IRanges(start=human_coord, end=human_coord), sample_name=sample_name, contig=contig))

exon_nearest_insertion = ref_annot_exons[nearest(gr_insertions, gr_exons, ignore.strand=T),]

```









```{r}

insertions_w_nearest_exon = bind_cols(data, tibble(exon_nearest_insertion) %>% rename(exon_lend = lend, exon_rend = rend))

insertions_w_nearest_exon = insertions_w_nearest_exon %>% 
  select(sample_name, splice_type, virus, virus_coord, humanchr, human_coord, gene_info, exon_lend, exon_rend) %>% rowwise() %>%
  mutate(human_delta = min(abs(human_coord - exon_lend), abs(human_coord - exon_rend) ) )



```


```{r}

# retrieve just the breakpoints involving splicing

spliced_insertions = insertions_w_nearest_exon %>% filter(splice_type != ".") %>% arrange(human_delta, gene_info, sample_name)

spliced_insertions
```

```{r}
gr_virus_spliceinfo = with(virus_splicing_info, GRanges(seqnames = virus, ranges=IRanges(start=coord, end=coord), virus_splice=splice ))


```

```{r}
gr_spliced_insertions = with(spliced_insertions, GRanges(seqnames=virus, ranges=IRanges(start=virus_coord, end=virus_coord)))

```


```{r}


virus_splice_nearest_virus_brkpt_insertions = virus_splicing_info[nearest(gr_spliced_insertions, gr_virus_spliceinfo, ignore.strand=T),]

```
```{r}

spliced_insertions = bind_cols(spliced_insertions, virus_splice_nearest_virus_brkpt_insertions %>% select(-virus) ) %>% 
  mutate(virus_delta = abs(coord - virus_coord))

```

```{r}

spliced_insertions = spliced_insertions %>% arrange(human_delta, virus_delta) %>% mutate(virus_spliced_human_refgene = (human_delta==1 & virus_delta==1) )

spliced_insertions

```

```{r}

write.table(spliced_insertions, "TCGA.spliced_viral_insertions.tsv", quote=F, sep="\t", row.names = F)


```


