---
title: "AnalyzeHotspots"
author: "bhaas"
date: '2023-03-27'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(cowplot)

```


# parse inputs

```{r}
HOTSPOT_WINSIZE = "1e5"
```


```{r}

hotspot_info = read.table( paste0("plot.hotspots.win_", HOTSPOT_WINSIZE, ".tsv.w_20_neighbors.regrouped_by_insert_gene.hotspot_virus_sample_counts.tsv"), header=T, sep="\t", stringsAsFactors = F)
# ex. plot.hotspots.win_1e5.tsv.w_20_neighbors.regrouped_by_insert_gene.hotspot_virus_sample_counts.tsv

head(hotspot_info)

```

```{r}

hotspot_insertions = read.table(paste0("hotspots.win_", HOTSPOT_WINSIZE, ".tsv.w_20_neighbors.regrouped_by_insert_gene"), 
                                header=T, sep="\t", stringsAsFactors = F)
# ex. hotspots.win_1e5.tsv.w_20_neighbors.regrouped_by_insert_gene

# use the regrouped hotspot name as the hotspot name:

hotspot_insertions = hotspot_insertions %>% mutate(hotspot = gene_regrouped_hotspot_name)

head(hotspot_insertions)
```


```{r}

#insertion_hotspot_assignments = read.table(paste0("hotspots.win_", HOTSPOT_WINSIZE, ".tsv"), header=T, sep="\t", stringsAsFactors = F)

#head(insertion_hotspot_assignments)
```



# begin analysis


```{r}

hotspot_info %>% select(hotspot) %>% unique() %>% nrow()

```

```{r}

hotspot_sum_sample_virus_counts = hotspot_info %>% select(hotspot, virus_genome, hotspot_virus_tally) %>% group_by(hotspot) %>% 
         summarize(sum_sample_viruses = sum(hotspot_virus_tally)) %>% arrange(desc(sum_sample_viruses)) %>% ungroup()

hotspot_sum_sample_virus_counts

```


```{r}

hotspot_sum_sample_virus_counts %>% ggplot(aes(x=reorder(hotspot, -1*sum_sample_viruses), sum_sample_viruses)) + geom_col() +
  ggtitle("sum(virus,sample) ranked hotspots") +
   scale_y_continuous(trans='log10')


```

```{r}

hotspot_sum_sample_virus_counts  %>% group_by(sum_sample_viruses) %>% tally()


```

```{r}

hotspot_sum_sample_virus_counts  %>% group_by(sum_sample_viruses) %>% tally() %>%
  ggplot(aes(x=as.factor(sum_sample_viruses), y=n)) + geom_col() +
  ggtitle("Number of hotspots with number of sample/viruses at sites") +
  xlab("number of sample/viruses at site") + ylab("number of hotspots")

```

```{r}


p = hotspot_sum_sample_virus_counts  %>% group_by(sum_sample_viruses) %>% tally() %>%
  ggplot(aes(x=as.factor(sum_sample_viruses), y=n)) + geom_col() +
  ggtitle("Number of hotspots with number of sample/viruses at sites") +
  xlab("number of sample/viruses at site") + ylab("number of hotspots") +
  scale_y_continuous(trans='log10')

p
```

```{r}
# include random data:

random_hotspot_data = read.table(gzfile(paste0("data/rand_cluster_size_counts.win_",  HOTSPOT_WINSIZE, ".tsv.gz")), header=F, sep="\t")

colnames(random_hotspot_data) = c('trial', 'hotspot_size', 'num_hotspots')

random_hotspot_data = random_hotspot_data %>% filter(hotspot_size >= 3)


```

```{r}

p = p + geom_boxplot(data=random_hotspot_data, aes(x=as.factor(hotspot_size), y=num_hotspots), color='red')

p

```



```{r}


top_hotspots = hotspot_info %>% group_by(hotspot) %>% summarize(sum_insertions = sum(hotspot_virus_tally)) %>% ungroup()

top_hotspots %>% arrange(desc(sum_insertions))


```



```{r}

 hotspot_info =  left_join(hotspot_info, top_hotspots, by='hotspot')

```


```{r}

hotspot_info %>% filter(sum_insertions >= 10) %>% 
  ggplot(aes(x=reorder(hotspot, -1*sum_insertions), y=hotspot_virus_tally, fill=virus_genome)) + geom_col() +
   theme(axis.text.x = element_text(angle = 90, hjust = 1))
  


```



```{r}

hotspot_info %>% filter(sum_insertions >= 10) %>% 
  mutate(virus_genome = ifelse(grepl("HPV", virus_genome), "HPV", virus_genome)) %>%
  ggplot(aes(x=reorder(hotspot, -1*sum_insertions), y=hotspot_virus_tally, fill=virus_genome)) + geom_col() +
   theme(axis.text.x = element_text(angle = 90, hjust = 1))
  

```


# Focus on significant hotspot sizes


```{r}

MIN_HOTSPOT_SIZE = 11

hotspot_info_top = hotspot_info %>% filter(sum_insertions >= MIN_HOTSPOT_SIZE) 

ordered_hotspots = hotspot_info_top %>% arrange(desc(sum_insertions)) %>% select(hotspot) %>% unique() %>% pull(hotspot)

hotspot_info_top$hotspot = factor(hotspot_info_top$hotspot, levels=ordered_hotspots)

hotspot_info_top %>% 
  #mutate(virus_genome = ifelse(grepl("HPV", virus_genome), "HPV", virus_genome)) %>%
  ggplot(aes(x=hotspot, y=hotspot_virus_tally, fill=virus_genome)) + geom_col() +
   theme(axis.text.x = element_text(angle = 90, hjust = 1))




hotspot_by_virus_plot = hotspot_info_top %>% 
  mutate(virus_genome = ifelse(grepl("HPV", virus_genome), "HPV", virus_genome)) %>%
  ggplot(aes(x=hotspot, y=hotspot_virus_tally, fill=virus_genome)) + geom_col() +
   theme(axis.text.x = element_text(angle = 90, hjust = 1))

hotspot_by_virus_plot

```



```{r}
hotspots_of_interest = hotspot_info_top %>% select(hotspot) %>% unique() %>% pull(hotspot)



hotspot_insertion_seqtypes = left_join(hotspot_info_top %>% select(hotspot),
                               hotspot_insertions %>% filter(gene_regrouped_hotspot_name %in% hotspots_of_interest) %>% 
                                 select(hotspot = gene_regrouped_hotspot_name, seqtype),
                                 by=c('hotspot') )


hotspot_seqtype_frac_info = hotspot_insertion_seqtypes %>% group_by(hotspot, seqtype) %>% tally() %>% mutate(pct=prop.table(n)) 

hotspot_seqtype_frac_info
```
```{r}
hotspots_of_interest


```


```{r}

hotspot_seqtype_frac_info$hotspot = factor(hotspot_seqtype_frac_info$hotspot, levels=ordered_hotspots)


hotspot_by_seqtype_plot = hotspot_seqtype_frac_info %>% ggplot(aes(x=hotspot, y=pct, fill=seqtype)) + geom_col() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1))

 hotspot_by_seqtype_plot 
  
```
  
  
```{r}
#  cohort representation

hotspot_insertion_cohorts = left_join(hotspot_info_top %>% select(hotspot),
                               hotspot_insertions %>% filter(gene_regrouped_hotspot_name %in% hotspots_of_interest) %>% 
                                 select(hotspot = gene_regrouped_hotspot_name, cohort),
                               by='hotspot')


hotspot_cohort_frac_info = hotspot_insertion_cohorts %>% group_by(hotspot, cohort) %>% tally() %>% mutate(pct=prop.table(n)) 

hotspot_cohort_frac_info

  
```







```{r}


hotspot_cohort_frac_info$hotspot = factor(hotspot_cohort_frac_info$hotspot, levels=ordered_hotspots)


hotspot_cohort_frac_info %>% ggplot(aes(x=hotspot, y=pct, fill=cohort)) + geom_col() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1))



```




```{r}

#  cohort representation

hotspot_insertions = hotspot_insertions  %>% mutate(proj_name = ifelse(cohort=='TCGA', paste(cohort, project, sep="^"), cohort))


hotspot_insertion_projects = left_join(hotspot_info_top %>% select(hotspot, hotspot),
                               hotspot_insertions %>% filter(gene_regrouped_hotspot_name %in% hotspots_of_interest) %>% 
                                 select(hotspot=gene_regrouped_hotspot_name, proj_name),
                               by='hotspot')


hotspot_project_frac_info = hotspot_insertion_projects %>% group_by(hotspot, proj_name) %>% tally() %>% mutate(pct=prop.table(n)) 

hotspot_project_frac_info



```


```{r}


hotspot_project_frac_info$hotspot = factor(hotspot_project_frac_info$hotspot, levels=ordered_hotspots)


hotspot_by_project_plot = hotspot_project_frac_info %>% ggplot(aes(x=hotspot, y=pct, fill=proj_name)) + geom_col() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1))

hotspot_by_project_plot

```



```{r}


pg = plot_grid(
          hotspot_by_seqtype_plot + theme(axis.title.x=element_blank(),axis.text.x=element_blank(), axis.ticks.x=element_blank()), 
          hotspot_by_project_plot + theme(axis.title.x=element_blank(),axis.text.x=element_blank(), axis.ticks.x=element_blank()), 
          hotspot_by_virus_plot, 
          ncol = 1, align='v',
          rel_heights=c(0.15, 0.15, 0.5))

pg
```

```{r}

ggsave(pg, file="top_hotspots_by_attributes.pdf", width=12, height=15)

```


# Incorporate expression and CNV info



```{r}
hotspot_info_top_with_insertions = left_join(hotspot_info_top %>% select(hotspot) %>% unique(),
                                             hotspot_insertions %>% select(sample_id, contig, hotspot),
                                             by='hotspot'
                                             )

nrow(hotspot_info_top_with_insertions)
```

```{r}
# add insertion cnv info

cnv_info = read.table("../Insertion_and_CNVs/TCGA_insertions_and_CNV_within10kb", header=T, sep="\t", stringsAsFactors = F)  %>% unique()

```

```{r}

hotspot_info_top_with_insertions_and_CNV = left_join(hotspot_info_top_with_insertions, 
                                                     cnv_info,
                                                     by=c('sample_id'='sample', 'contig'='insertion') )

nrow(hotspot_info_top_with_insertions_and_CNV)

```


```{r}

# choose peak cnv copy number where multiple insertions analyzed for a single sample

hotspot_info_top_with_insertions_and_CNV = hotspot_info_top_with_insertions_and_CNV  %>% group_by(hotspot, sample_id) %>% arrange(desc(copy_number)) %>% filter(row_number()==1) %>% ungroup()


nrow(hotspot_info_top_with_insertions_and_CNV)
 
```

```{r}

hotspot_info_top_with_insertions_and_CNV$hotspot = factor(hotspot_info_top_with_insertions_and_CNV$hotspot, levels=ordered_hotspots)

```


```{r}

hotspot_cnv_boxplot = hotspot_info_top_with_insertions_and_CNV %>% ggplot(aes(x=hotspot, y=copy_number)) + geom_boxplot() +
  geom_hline(yintercept=2, linetype=2, color='purple') +
   theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  scale_y_continuous(trans='log10')

hotspot_cnv_boxplot 

```

# Incorporate expression effects

```{r}

insertion_expr_info = read.table("../Insertion_and_EXPR_effects/virus_insertions_unfiltered.10kb.expression_region_analysis.tsv", 
                                 header=T, sep="\t", stringsAsFactors = F) %>% unique()


nrow(insertion_expr_info)
```






```{r}

hotspot_info_top_with_insertions_and_CNV_and_EXPR = left_join(hotspot_info_top_with_insertions_and_CNV, 
                                                              insertion_expr_info %>% select(sample_id, contig, expr_quantile),
                                                              by=c('sample_id', 'contig'))

nrow(hotspot_info_top_with_insertions_and_CNV_and_EXPR)
```

```{r}

# take single highest expr quantile for sample,hotspot combo
hotspot_info_top_with_insertions_and_CNV_and_EXPR = hotspot_info_top_with_insertions_and_CNV_and_EXPR %>%
  group_by(hotspot, sample_id) %>% arrange(desc(expr_quantile)) %>% filter(row_number()==1) %>% ungroup()

nrow(hotspot_info_top_with_insertions_and_CNV_and_EXPR)

```



```{r}

hotspot_info_top_with_insertions_and_CNV_and_EXPR$hotspot = factor(hotspot_info_top_with_insertions_and_CNV_and_EXPR$hotspot, levels=ordered_hotspots)

```


```{r}

hotspot_expr_quantile_plot = hotspot_info_top_with_insertions_and_CNV_and_EXPR %>% ggplot(aes(x=hotspot, y=expr_quantile)) + geom_boxplot() +
   theme(axis.text.x = element_text(angle = 90, hjust = 1))

hotspot_expr_quantile_plot
```



Interesting:  CESC-LP-A5U3-TP  has a couple of hotspots and expr effects:
 
chr17:39735144^ERBB2  19.7 cnv and 0.98 expr quantile
chr8:127790761^MYC~PVT1_1  3 cnv and 0.57 expr quantile


What other samples have multiple hotspot insertions?

```{r}
sample_ids_mult_hotspots = hotspot_info_top_with_insertions_and_CNV_and_EXPR %>% select(hotspot, sample_id) %>%  unique() %>%
  group_by(sample_id) %>% tally() %>% filter(n>1) %>% ungroup()
  
sample_ids_mult_hotspots
```

```{r}

hotspot_info_top_with_insertions_and_CNV_and_EXPR %>% filter(sample_id %in% sample_ids_mult_hotspots$sample_id) %>%
  select(sample_id, hotspot, copy_number, expr_quantile) %>% arrange(sample_id)

```



```{r}
# Bring all the plots together.

pg2 = plot_grid(
          hotspot_by_seqtype_plot + theme(axis.title.x=element_blank(),axis.text.x=element_blank(), axis.ticks.x=element_blank()), 
          hotspot_by_project_plot + theme(axis.title.x=element_blank(),axis.text.x=element_blank(), axis.ticks.x=element_blank()), 
          hotspot_cnv_boxplot + theme(axis.title.x=element_blank(),axis.text.x=element_blank(), axis.ticks.x=element_blank()),
          hotspot_expr_quantile_plot + theme(axis.title.x=element_blank(),axis.text.x=element_blank(), axis.ticks.x=element_blank()),
          hotspot_by_virus_plot, 
          ncol = 1, 
          align='v',
          axis='lr',
          rel_heights=c(0.15, 0.15, 
                        0.1,
                        0.1, 
                        0.5
                        ))

pg2
```

```{r}

ggsave(pg2, file="hotspot_summary_plot.pdf", width=20, height=15)



```




```{r}
# Bring all the plots together.

pg3 = plot_grid(
          #hotspot_by_seqtype_plot + theme(axis.title.x=element_blank(),axis.text.x=element_blank(), axis.ticks.x=element_blank()), 
          #hotspot_by_project_plot + theme(axis.title.x=element_blank(),axis.text.x=element_blank(), axis.ticks.x=element_blank()), 
          hotspot_cnv_boxplot + theme(axis.title.x=element_blank(),axis.text.x=element_blank(), axis.ticks.x=element_blank()),
          hotspot_expr_quantile_plot, # + theme(axis.title.x=element_blank(),axis.text.x=element_blank(), axis.ticks.x=element_blank()),
          #hotspot_by_virus_plot, 
          ncol = 1, align='v',
          rel_heights=c(#0.15, 0.15, 
                        0.2,
                        0.7#, 
                        #0.5
                        ))

pg3
```


# Compare sample representation among hotspots

```{r}

ordered_hotspot_insertions = hotspot_insertions %>% 
  select(hotspot, hotspot_sample_counts_redef, hotspot_sample_counts_redef_list) %>% unique() 

ordered_hotspot_insertions$hotspot_chromosome = sapply(ordered_hotspot_insertions$hotspot, function(x) {
  str_split(x, ":")[[1]][1]
} )

ordered_hotspot_insertions$hotspot_coord = sapply(ordered_hotspot_insertions$hotspot, function(x) {
  as.numeric(str_split(x, ":|\\^")[[1]][2])
} )

ordered_hotspot_insertions = ordered_hotspot_insertions %>% arrange(hotspot_chromosome, hotspot_coord)

n = nrow(ordered_hotspot_insertions)
lookback = 10

```


```{r}

# peak at chr8 hotspot

# ordered_hotspot_insertions %>% filter(grepl("^chr8:", hotspot)) %>% view()


```


```{r}

# Are there similar sample compositions for neighboring hotspots suggesting that they should be combined into single hotspots?

MIN_J = 0.5

find_related_hotspots = function(x) {
  
  jaccard_pair_df = NULL
  
  index_row = ordered_hotspot_insertions[x,]
  index_row_sample_list = str_split(index_row$hotspot_sample_counts_redef_list, ",")
  
  
  for (i in seq(x-lookback+1, x-1)) {
    compare_row = ordered_hotspot_insertions[i,]
    compare_row_sample_list = str_split(compare_row, ",")
    
    u = union(index_row_sample_list, compare_row_sample_list)
    insect = intersect(index_row_sample_list, compare_row_sample_list)
    j = length(insect)/length(u)
    if (j >= MIN_J) {
      jaccard_pair_df = bind_rows(jaccard_pair_df, 
                                  data.frame(hotspot_A = index_row$hotspot, hotspot_B = compare_row$hotspot, j=j) )
    }
    
  }
  
  jaccard_pair_df
}

hotspot_sample_jaccard = do.call(bind_rows, lapply(seq(lookback, n, 1), find_related_hotspots))

#lapply(seq(lookback, 20, 1), find_related_hotspots)

hotspot_sample_jaccard

# nope

```


# try matrix/heatmap format instead of barchart.

```{r}

p_hotspot_project_frac_info = hotspot_project_frac_info  %>% ggplot(aes(x=hotspot, y=proj_name, fill=pct)) + geom_tile() +
  scale_fill_gradient(low="lightblue", high="black", na.value = 'white') 

p_hotspot_project_frac_info

```

```{r}


hotspot_info_top %>% ggplot(aes(x=hotspot, y=virus_genome, fill=hotspot_virus_tally)) + geom_tile()  +
  scale_fill_gradient(low="lightblue", high="black", na.value = 'white') 


```




```{r}

p_hotspot_virus_frac_rep = hotspot_info_top %>% mutate(frac_insertions = hotspot_virus_tally/sum_insertions) %>%
  ggplot(aes(x=hotspot, y=virus_genome, fill=frac_insertions)) + geom_tile()  +
  scale_fill_gradient(low="lightblue", high="black", na.value = 'white') 


p_hotspot_virus_frac_rep

```


```{r}
# Bring all the plots together.

pg3 = plot_grid(
          hotspot_by_seqtype_plot + theme(axis.title.x=element_blank(),axis.text.x=element_blank(), axis.ticks.x=element_blank()), 
          p_hotspot_project_frac_info  + theme(axis.title.x=element_blank(),axis.text.x=element_blank(), axis.ticks.x=element_blank()), 
          hotspot_cnv_boxplot + theme(axis.title.x=element_blank(),axis.text.x=element_blank(), axis.ticks.x=element_blank()),
          hotspot_expr_quantile_plot + theme(axis.title.x=element_blank(),axis.text.x=element_blank(), axis.ticks.x=element_blank()),
          p_hotspot_virus_frac_rep +  theme(axis.text.x = element_text(angle = 90, hjust = 1)),
          ncol = 1, 
          align='v',
          axis='lr',
          rel_heights=c(0.15, 0.15, 
                        0.1,
                        0.1, 
                        0.5
                        ))

pg3


```



# examine virus representation among the different project cohorts

```{r}

top_hotspots = hotspot_info_top %>% select(hotspot) %>% unique() %>% pull(hotspot)


virus_project_hotspot_rep = hotspot_insertions %>% 
  filter(hotspot %in% top_hotspots) %>%
  mutate(proj_name = ifelse(cohort=='GTEx', paste(cohort, project), proj_name)) %>%
  select(hotspot, virus_genome, cohort, proj_name, participant) %>% unique() %>%
    group_by(hotspot, virus_genome, cohort, proj_name) %>% tally(name='num_participants') %>% ungroup()

virus_project_hotspot_rep$hotspot = factor(virus_project_hotspot_rep$hotspot, levels=ordered_hotspots)


```


```{r}

hotspots_in_plot = virus_project_hotspot_rep %>% select(hotspot) %>% unique()

```


```{r}


ccle_virus_data = left_join(hotspots_in_plot, 
                                virus_project_hotspot_rep %>% filter(cohort=='CCLE') 
                                )  
  
ccle_virus_data$hotspot = factor(ccle_virus_data$hotspot, levels=ordered_hotspots)

ccle_virus_rep_plot =   ccle_virus_data %>%  ggplot(aes(x=hotspot, y=virus_genome, fill=num_participants)) + geom_tile() + ggtitle("CCLE")  +
  scale_fill_gradient(low="lightblue", high="black", na.value = 'white') 

ccle_virus_rep_plot + theme(axis.text.x = element_text(angle = 90, hjust = 1))
```

```{r}

tcga_virus_data = left_join(hotspots_in_plot, 
                                virus_project_hotspot_rep %>% filter(cohort=='TCGA') ) 

tcga_virus_data$hotspot = factor(tcga_virus_data$hotspot, levels=ordered_hotspots)

tcga_virus_rep_plot = tcga_virus_data %>% 
  ggplot(aes(x=hotspot, y=virus_genome, fill=num_participants)) + geom_tile() + ggtitle("TCGA")  +
  scale_fill_gradient(low="lightblue", high="black", na.value = 'white') 

tcga_virus_rep_plot  + theme(axis.text.x = element_text(angle = 90, hjust = 1))


```





```{r}

gtex_virus_data  = left_join(hotspots_in_plot, 
                                 virus_project_hotspot_rep %>% filter(cohort=='GTEx') ) %>% 
  mutate(virus_projname = paste(virus_genome, proj_name)) 

gtex_virus_data$hotspot = factor(gtex_virus_data$hotspot, levels=ordered_hotspots)

gtex_virus_rep_plot =  gtex_virus_data %>%  ggplot(aes(x=hotspot, y=virus_projname, fill=num_participants)) + geom_tile() + ggtitle("GTEx")  +
  scale_fill_gradient(low="lightblue", high="black", na.value = 'white') 

gtex_virus_rep_plot +  theme(axis.text.x = element_text(angle = 90, hjust = 1))

```





```{r}

natgen_virus_data = left_join(hotspots_in_plot, 
                                  virus_project_hotspot_rep %>% filter(cohort=='NATGEN2015') ) 


natgen_virus_data$hotspot = factor(natgen_virus_data$hotspot, levels=ordered_hotspots)

natgen_virus_rep_plot = natgen_virus_data %>% ggplot(aes(x=hotspot, y=virus_genome, fill=num_participants)) + geom_tile() + ggtitle("NATGEN2015")  +
  scale_fill_gradient(low="lightblue", high="black", na.value = 'white') 


natgen_virus_rep_plot +  theme(axis.text.x = element_text(angle = 90, hjust = 1))

```

```{r}

pg4 = plot_grid(
          gtex_virus_rep_plot  + theme(axis.title.x=element_blank(),axis.text.x=element_blank(), axis.ticks.x=element_blank()), 
          ccle_virus_rep_plot + theme(axis.title.x=element_blank(),axis.text.x=element_blank(), axis.ticks.x=element_blank()),
           natgen_virus_rep_plot + theme(axis.title.x=element_blank(),axis.text.x=element_blank(), axis.ticks.x=element_blank()), 
          tcga_virus_rep_plot +  theme(axis.text.x = element_text(angle = 90, hjust = 1)),
          ncol = 1, 
          align='v',
          axis='lr',
          rel_heights=c(0.1,
                        0.1,
                        0.1, 
                        0.3
                        ))

pg4

```

```{r}

ggsave(pg4, file="virus_cohort_representation.pdf", width=25, height=15)


```



# Examine participant and virus representation according to cohort.

```{r}

natgen2015_sample_virus_info = hotspot_insertions %>% filter(cohort=="NATGEN2015") %>% 
   group_by(sample_id, virus) %>% summarize(max_total_rpm = max(total_rpm)) %>%
    select(sample_id, virus, max_total_rpm) %>% unique()

natgen2015_sample_virus_info %>%
     ggplot(aes(x=sample_id, y=virus, fill=log10(max_total_rpm+1))) + geom_tile() +
     scale_fill_gradient(low="lightblue", high="black", na.value = 'white') +
    theme(axis.text.x = element_text(angle = 90, hjust = 1))


```
```{r}

# heatmap for above virus/sample associations, clustered.
natgen2015_sample_virus_info_matrix = natgen2015_sample_virus_info %>% spread(key=virus, value=max_total_rpm, fill=0)

write.table(natgen2015_sample_virus_info_matrix, file="natgen2015_sample_virus_info.matrix", quote=F, sep="\t", row.names=F)

# ~/GITHUB/trinityrnaseq/Analysis/DifferentialExpression/PtR  -m natgen2015_sample_virus_info.matrix --heatmap --log2 --heatmap_colorscheme "white,blue" --heatmap_scale_limits "0,4"

```

