---
title: "InsertionFusionMapping.Rmd"
author: "bhaas"
date: '2023-04-05'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
```


```{r}

insertions_data = read.table("../virus_insertions_unfiltered.tsv", header=T, sep="\t", stringsAsFactors = F)
 
orig_insertions_data = insertions_data # for later

insertion_samples = insertions_data %>% select(sample_id) %>% unique() %>% pull(sample_id)

```

```{r}

starF_fusions_data = read.table(gzfile("../../data/TCGA_n_GTEx.STAR-Fusion.v1.7.fusion_preds.tsv.gz"), header=T, sep="\t", stringsAsFactors = F)

# restrict to samples with insertions

fusion_samples = starF_fusions_data %>% select(sample) %>% unique() %>% pull(sample)

```


```{r}

fusion_samples_with_insertions = intersect(insertion_samples, fusion_samples)

length(fusion_samples_with_insertions)

# 637 samples with insertions and fusions

```



```{r}

insertions_data = insertions_data %>% filter(sample_id %in% fusion_samples_with_insertions)

starF_fusions_data = starF_fusions_data  %>% filter(sample %in% fusion_samples_with_insertions)


```


```{r}

insertions_data = insertions_data %>% select(sample_id, contig, humanchr, human_coord) %>% unique()

```


```{r}

starF_fusions_coords = bind_rows(starF_fusions_data  %>% select(sample, fusion_name, breakpoint=LeftBreakpoint),
                                 starF_fusions_data %>% select(sample, fusion_name, breakpoint=RightBreakpoint) ) %>% unique()

```

```{r}

starF_fusions_coords = starF_fusions_coords %>% rowwise() %>% 
  mutate(Chromosome = str_split(breakpoint, ":")[[1]][1] ) %>%
  mutate(BreakpointCoord = str_split(breakpoint, ":")[[1]][2] )

starF_fusions_coords$BreakpointCoord = as.numeric(starF_fusions_coords$BreakpointCoord )

```

```{r}

library(GenomicRanges)

df_A_with_nearest_df_B = function(df_A, df_B) {
  
    # each df requires fields:  Chromosome_val, Start_val, End_val
  
    gr_df_A = with(df_A, GRanges(Chromosome_val,
                                IRanges(start=Start_val, end=End_val)))
    
    gr_df_B = with(df_B, GRanges(Chromosome_val,
                                IRanges(start=Start_val, end=End_val)))

  
    nearest_df_B = df_B[nearest(gr_df_A, gr_df_B, ignore.strand=T),]
    
    df_A_with_nearest_df_B = bind_cols(df_A %>% select(-Chromosome_val, -Start_val, -End_val), 
                                       nearest_df_B %>% select(-Chromosome_val, -Start_val, -End_val))
    
    return(df_A_with_nearest_df_B)

}


```


```{r}

insertions_data  = insertions_data  %>% mutate(Chromosome_val = humanchr, Start_val = human_coord, End_val = human_coord)

starF_fusions_coords = starF_fusions_coords %>% mutate(Chromosome_val = Chromosome, Start_val = BreakpointCoord, End_val = BreakpointCoord)

```


```{r}

fusion_mapped_insertions = NULL

MAX_DIST_INSERTION_TO_BREAKPOINT = 100000

for (sample_select in fusion_samples_with_insertions) {
  
    starF_fusions_coords_sample = starF_fusions_coords %>% filter(sample == sample_select)
    insertions_data_sample = insertions_data %>% filter(sample_id == sample_select)
    
    fusion_sample_mapped_insertions = df_A_with_nearest_df_B(insertions_data_sample, starF_fusions_coords_sample)
      
    fusion_sample_mapped_insertions = fusion_sample_mapped_insertions %>% filter(! is.na(BreakpointCoord)) %>%
      mutate(delta = abs(BreakpointCoord - human_coord)) %>%
      filter(delta <= MAX_DIST_INSERTION_TO_BREAKPOINT)
    
    if (nrow(fusion_sample_mapped_insertions > 0)) {
        fusion_mapped_insertions = bind_rows(fusion_mapped_insertions, fusion_sample_mapped_insertions)
    }
    
}


fusion_mapped_insertions
```


```{r}

fusion_mapped_insertions = fusion_mapped_insertions %>% select(sample_id, contig, humanchr, human_coord, fusion_name, fusionBreakpointCoord = BreakpointCoord, DistBetweenInsertionAndFusionBreakpoint = delta)

write.table(fusion_mapped_insertions, "fusions_at_insertions.tsv", sep="\t", quote=F, row.names=F)


```


```{r}
# integrate fusions info into original virus insertion table

fusion_mapped_insertions = fusion_mapped_insertions %>% select(sample_id, contig, fusion_at_insertion = fusion_name) %>% unique()

orig_insertions_data = left_join(orig_insertions_data, fusion_mapped_insertions, by=c('sample_id', 'contig'))

orig_insertions_data %>% filter(! is.na(fusion_at_insertion))
```


```{r}

write.table(orig_insertions_data, file="../virus_insertions_unfiltered.w_fusions.tsv", sep="\t", quote=F, row.names=F)

```


# Examine fusion/virus associations

```{r}

fusion_sample_virus_counts = orig_insertions_data %>% select(sample_name, virus, fusion_at_insertion) %>% filter(! is.na(fusion_at_insertion)) %>% 
  unique() %>%
  group_by(fusion_at_insertion) %>% tally(name='fusion_sample_virus_count') %>% arrange(desc(fusion_sample_virus_count))


fusion_sample_virus_counts
```

```{r}

fusions_at_insertions_min2 = fusion_sample_virus_counts %>% filter(fusion_sample_virus_count >= 2) %>% pull(fusion_at_insertion)

```


```{r}

insertions_w_fusions_min2 = orig_insertions_data  %>% 
  filter(fusion_at_insertion %in% fusions_at_insertions_min2) %>% 
  select(sample_name, humanchr, human_coord, virus, fusion_at_insertion, total, total_rpm) %>%
  group_by(sample_name, humanchr, virus, fusion_at_insertion) %>%
  arrange(desc(total_rpm)) %>% filter(row_number()==1) %>% ungroup()

insertions_w_fusions_min2

```

```{r}

insertions_w_fusions_min2  %>% group_by(virus, fusion_at_insertion) %>% tally(name='sample_count') %>%
  ggplot(aes(x=reorder(fusion_at_insertion, -1*sample_count), y=sample_count, fill=virus)) + geom_col() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1))



# PVT1::MYC in 5 samples
# MIR1204::MYC in 2 samples
# GRB7::ERBB2 in 2 samples  - known oncogenic fusion
# STK25::IGH for Macacine alphaherpesvirus_1  - looks like trace levels
# NCOR::UBC - found in normals

```

```{r}

insertions_w_fusions_min2 %>% filter(! grepl("Eppstein", virus)) %>% arrange(fusion_at_insertion, desc(total_rpm))

```



