---
title: "compare_VIF_to_published_NatGenet2015"
author: "bhaas"
date: '2023-03-01'
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(GenomicRanges)
```


```{r}

BINSIZE = 1000

```

# published insertions

```{r}

## Just HIVID here

published_insertions = read.table("../../NatGenetics2015_reanalysis_published_data/data/TableS5.w_Hg38.tsv", header=T, sep="\t", stringsAsFactors = F)  %>%
  select(sample_id, hg38_Chrom, hg38_Coord, pub_total = total, found_by_WGS )

published_insertions = published_insertions %>% mutate(sample_id = toupper(sample_id)) %>% 
  mutate(sample_id = ifelse(sample_id == "CASKI", "CASKID", sample_id))

nrow(published_insertions)

# in the abstract it says 3667 and in the paper 3666.
# there are actually 

```


```{r}
# num samples
published_insertions %>% select(sample_id) %>% unique() %>% nrow()

# 103/135 hivid samples have insertions, as indicated in the paper

```

```{r}

# There are three insertions that we couldn't get the hg19->hg38 coordinate translation for using liftover
published_insertions %>% filter(is.na(hg38_Coord))

published_insertions = published_insertions %>% filter(! is.na(hg38_Coord))

```



```{r}

# how may insertion loci - taking the highest supported insertion within the BINSIZE?

published_insertions = published_insertions %>% mutate(bin = round(hg38_Coord / BINSIZE)) %>% 
  group_by(sample_id, hg38_Chrom, bin) %>% arrange(desc(pub_total)) %>% filter(row_number() == 1) %>% ungroup() %>% 
  select(-bin)

nrow(published_insertions)

```


```{r}
# What's the distribution of read support across loci?

median(published_insertions$pub_total)

published_insertions %>% group_by(pub_total) %>% tally() %>% ggplot(aes(x=pub_total, y=n)) + geom_col() + xlim(0,25)

# median of 5 reads for support at locus

```


```{r}

# insertions per sample

published_insertions %>% group_by(sample_id) %>% tally() %>% 
  ggplot(aes(x=reorder(sample_id, -1*n), n)) + geom_col() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) + 
  ggtitle("Published Insertions/Sample")


```


# VIF insertions

```{r}

vif_insertions = read.table("../insertions.tsv", header=T, sep="\t", stringsAsFactors = F) %>% filter(sample_type == "hivid")

vif_insertions$sample_id = vif_insertions$sample_label_short

nrow(vif_insertions)

```


```{r}
# identify the samples we were able to process.

vif_processed = read.table(gzfile("../data/NatGen2015.star_align_stats.tsv.gz"), header=T, sep="\t", stringsAsFactors = F)

sample_info = read.table("../NatGen2015.samples_info.tsv", header=T, sep="\t", stringsAsFactors = F) %>% 
  filter(sample_type == "hivid") %>%
  mutate(sample_id = sample_label_short)

sample_info = sample_info %>% mutate(vif_processed = TRUE)

sample_info = full_join(sample_info, published_insertions %>% select(sample_id) %>% unique(), by='sample_id' )


sample_info = sample_info %>% mutate(vif_processed = ifelse(is.na(vif_processed), FALSE, vif_processed))
    
sample_info %>% arrange(vif_processed)                         
                                     
# we don't have S12 hivid data!  Can't find it on SRA

```

```{r}

vif_hivid_samples_processed = sample_info %>% filter(vif_processed) %>% pull(sample_id)


```


```{r}
# convert to insertion loci
vif_insertions = vif_insertions %>% 
  filter(is_primary == 'True') %>%  # restrict to single virus breakends
  mutate(bin = round(human_coord / BINSIZE) ) %>% 
  group_by(sample_id, humanchr, bin) %>% arrange(desc(total)) %>% filter(row_number() == 1) %>% ungroup() %>% 
  select(-bin)

nrow(vif_insertions)


```



```{r}

# vif insertions per sample


vif_insertions %>% group_by(sample_id) %>% tally() %>% 
  ggplot(aes(x=reorder(sample_id, -1*n), n)) + geom_col() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) + 
  ggtitle("VIF Insertions/Sample")


```






```{r}
# merge published and VIF counts into single table

combined_insertion_counts = full_join(vif_insertions %>% group_by(sample_id) %>% tally(name='vif_insertion_count'),
          published_insertions %>% 
            filter(sample_id %in% vif_hivid_samples_processed) %>% 
            group_by(sample_id) %>% tally(name='published_insertion_count'),
          by='sample_id',
          )

combined_insertion_counts = combined_insertion_counts %>%
  mutate(vif_insertion_count = ifelse(is.na(vif_insertion_count), 0, vif_insertion_count)) %>%
  mutate(published_insertion_count = ifelse(is.na(published_insertion_count), 0, published_insertion_count))

combined_insertion_counts %>% arrange(desc(vif_insertion_count))

```

```{r}

combined_insertion_counts %>% ggplot(aes(y=vif_insertion_count+1, x=published_insertion_count+1)) + geom_point() +
  scale_x_continuous(trans='log10') +
  scale_y_continuous(trans='log10') + geom_abline(slope=1, intercept=0, color='red') +
  ggtitle("Comparison of # Insertions per Sample") +
  xlab("NatGen2015") + ylab("ctat-VIF")
  

```

```{r}

cor.test(combined_insertion_counts$vif_insertion_count, combined_insertion_counts$published_insertion_count)


```

```{r}

combined_insertion_counts %>% summarize(sum_vif_insertions = sum(vif_insertion_count), sum_pub_insertions = sum(published_insertion_count))


```


```{r}

# what if we remove those samples that we find insertions in, but publication says none?

combined_insertion_counts %>% filter(published_insertion_count > 0) %>%
  summarize(sum_vif_insertions = sum(vif_insertion_count), sum_pub_insertions = sum(published_insertion_count))


```







```{r}

sample_ids = combined_insertion_counts %>% filter(published_insertion_count > 0 & vif_insertion_count > 0) %>% pull(sample_id)

length(sample_ids)


```


```{r}

df_A_with_nearest_df_B = function(df_A, df_B) {
  
    # each df requires fields:  Chromosome_val, Start_val, End_val
  
    gr_df_A = with(df_A, GRanges(Chromosome_val,
                                IRanges(start=Start_val, end=End_val)))
    
    gr_df_B = with(df_B, GRanges(Chromosome_val,
                                IRanges(start=Start_val, end=End_val)))

  
    nearest_df_B = df_B[nearest(gr_df_A, gr_df_B, ignore.strand=T),]
    
    df_A_with_nearest_df_B = bind_cols(df_A %>% select(-Chromosome_val, -Start_val, -End_val), 
                                       nearest_df_B %>% select(-Chromosome_val, -Start_val, -End_val))
    
    return(df_A_with_nearest_df_B)

}



```






```{r}

vif_with_nearest_published = NULL

published_with_nearest_vif = NULL

for (selected_sample_id in sample_ids) {
  
  published_insertions_for_sample = published_insertions %>% filter(sample_id == selected_sample_id) %>%
    mutate(Chromosome_val = hg38_Chrom, Start_val = hg38_Coord, End_val = hg38_Coord)
  
  vif_insertions_for_sample = vif_insertions %>% filter(sample_id == selected_sample_id) %>% select(-sample_id) %>%
    mutate(Chromosome_val = humanchr, Start_val = human_coord, End_val = human_coord)
  

  vif_with_nearest_published = bind_rows(vif_with_nearest_published, 
                                            df_A_with_nearest_df_B(vif_insertions_for_sample, published_insertions_for_sample) )
  
  
  published_with_nearest_vif = bind_rows(published_with_nearest_vif,
                                            df_A_with_nearest_df_B(published_insertions_for_sample, vif_insertions_for_sample))
  
}


```



```{r}

vif_with_nearest_published = vif_with_nearest_published %>% mutate(delta = abs(hg38_Coord - human_coord))


published_with_nearest_vif = published_with_nearest_vif %>% mutate(delta = abs(hg38_Coord - human_coord))

```


```{r}

AGREE_DIST_RANGE = BINSIZE

vif_with_nearest_published  = vif_with_nearest_published  %>% mutate(within_range = delta <= AGREE_DIST_RANGE)

published_with_nearest_vif  = published_with_nearest_vif  %>% mutate(within_range = delta <= AGREE_DIST_RANGE)


```


```{r}
## Examine fraction of vif insertions that agree

min_read_supports = c(1, 2, 5, 10, 20, 50) 

vif_with_nearest_published_min_support_df = NULL

for (min_read_support in min_read_supports) {
  
  vif_with_nearest_published_min_reads = vif_with_nearest_published  %>% filter(total >= min_read_support)
  
  num_insertions = nrow(vif_with_nearest_published_min_reads)
  num_agree = nrow(vif_with_nearest_published_min_reads %>% filter(within_range))
  frac_agree = num_agree / num_insertions
  
  vif_with_nearest_published_min_support_df = bind_rows(vif_with_nearest_published_min_support_df, 
                                                           data.frame(min_read_support=min_read_support, 
                                                                      num_insertions = num_insertions,
                                                                      num_agree = num_agree,
                                                                      frac_agree = frac_agree))
  
}
  

vif_with_nearest_published_min_support_df

```


```{r}


published_with_nearest_vif_min_support_df = NULL

for (min_read_support in min_read_supports) {
  
  published_with_nearest_vif_min_reads = published_with_nearest_vif  %>% filter(pub_total >= min_read_support)
  
  num_insertions = nrow(published_with_nearest_vif_min_reads)
  num_agree = nrow(published_with_nearest_vif_min_reads %>% filter(within_range))
  frac_agree = num_agree / num_insertions
  
  published_with_nearest_vif_min_support_df = bind_rows(published_with_nearest_vif_min_support_df, 
                                                           data.frame(min_read_support=min_read_support, 
                                                                      num_insertions = num_insertions,
                                                                      num_agree = num_agree,
                                                                      frac_agree = frac_agree))
  
}
  

published_with_nearest_vif_min_support_df


```

```{r}


published_with_nearest_vif %>% select(sample_id, humanchr, hg38_Coord, human_coord, delta, within_range, pub_total) %>%
  arrange(sample_id, humanchr, delta) %>% filter(pub_total >= 10)


```



```{r}

# for those insertion sites that agree, compare the read support found

published_with_nearest_vif %>% 
  filter(within_range) %>%
  ggplot(aes(x=pub_total, y=total)) + geom_point() +
  geom_abline(slope=1, intercept=0, color='red') +
  scale_x_continuous(trans='log10') +
  scale_y_continuous(trans='log10') + geom_abline(slope=1, intercept=0, color='red') +
  ggtitle("Read support for agreed-upon insertion sites - HIVID") +
  xlab("NatGen2015") +
  ylab("ctat-VIF")


```


```{r}

published_with_nearest_vif %>% 
  filter(within_range) %>%
  mutate(count_delta = total - pub_total) %>%
  ggplot() + 
  geom_boxplot(aes(x=count_delta))


```

```{r}

totals = published_with_nearest_vif %>% 
  filter(within_range) %>%
  select(total, pub_total)

cor.test(totals$total, totals$pub_total)


```


