---
title: "InsertionToGeneAndSplice"
author: "bhaas"
date: '2023-04-07'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
```



```{r}

insertions_data = read.table("../InsertionAnalysis/all_insertions_and_virus_content_merged.FILTERED.tsv", header=T, sep="\t", stringsAsFactors = F) %>%
  filter(seqtype=="RNA")

```


```{r}

exon_data = read.table(gzfile("../../data/ref_annot.gtf.exon_features.gz"), header=F, sep="\t", stringsAsFactors = F)
exon_data = exon_data[,c(1,4,5,9)]
colnames(exon_data) = c('Chromosome', 'Start', 'End', 'Gene_ID')
exon_data$Gene_ID = str_replace(exon_data$Gene_ID, "gene_id ", "")

```


```{r}

insertions_coords = insertions_data %>% select(humanchr, human_coord) %>% unique()


```

```{r}
library(GenomicRanges)

df_A_with_nearest_df_B = function(df_A, df_B) {
  
    # each df requires fields:  Chromosome_val, Start_val, End_val
  
    gr_df_A = with(df_A, GRanges(Chromosome_val,
                                IRanges(start=Start_val, end=End_val)))
    
    gr_df_B = with(df_B, GRanges(Chromosome_val,
                                IRanges(start=Start_val, end=End_val)))

  
    nearest_df_B = df_B[nearest(gr_df_A, gr_df_B, ignore.strand=T),]
    
    df_A_with_nearest_df_B = bind_cols(df_A %>% select(-Chromosome_val, -Start_val, -End_val), 
                                       nearest_df_B %>% select(-Chromosome_val, -Start_val, -End_val))
    
    return(df_A_with_nearest_df_B)

}
```


```{r}
insertions_coords = insertions_coords %>% mutate(Chromosome_val = humanchr, Start_val = human_coord, End_val = human_coord)

exon_data = bind_rows(exon_data %>% mutate(Chromosome_val = Chromosome, Start_val = Start, End_val = Start),
                      exon_data %>% mutate(Chromosome_val = Chromosome, Start_val = End, End_val = End)) %>%
  unique()

```

```{r}

insertions_coords_w_nearest_exon = df_A_with_nearest_df_B(insertions_coords, exon_data)


```

```{r}

insertions_coords_w_nearest_exon = insertions_coords_w_nearest_exon %>% rowwise() %>% 
  mutate(min_dist_from_exon_end = min( abs(Start-human_coord), abs(End-human_coord))) %>%
  mutate(exon_insert_type = ifelse(human_coord > Start & human_coord < End, "within_exon", "external")) %>%
  mutate(exon_insert_type = ifelse(min_dist_from_exon_end == 1, "at_ref_exon_boundary", exon_insert_type))


```

```{r}
head(insertions_coords_w_nearest_exon)

```

```{r}

write.table(insertions_coords_w_nearest_exon, file='insertions_mapped_to_gene_exons.tsv', sep="\t", quote=F, row.names = F)

```


```{r}

insertions_data = left_join(insertions_data,
                            insertions_coords_w_nearest_exon %>% mutate(exon_start = Start, exon_end = End) %>% select(-Chromosome, -Start, -End),
                            by=c('humanchr', 'human_coord') )


```


```{r}


insertions_data %>% filter(exon_insert_type == "at_ref_exon_boundary") %>%
  select(sample_id, virus, contig, splice_type, Gene_ID)



```


```{r}

# insertions that have both spliced and unspliced breakpoints at ref exon boundaries - curious about the unspliced ones and what they represent.

insertions_data %>% filter(exon_insert_type == "at_ref_exon_boundary") %>% 
  select(sample_id, Gene_ID, splice_type) %>% unique() %>%
  group_by(Gene_ID, splice_type) %>% tally() %>% 
  spread(key=splice_type, value=n)  %>%
  filter(! is.na(`.`) & ! is.na(`GT-AG`)) %>%
  arrange(desc(`GT-AG`))


```



```{r}

# Examine TERT results

insertions_data %>% filter(Gene_ID == "TERT^ENSG00000164362.17") %>% 
   filter(exon_insert_type == "at_ref_exon_boundary") %>%
  select(sample_name, humanchr, human_coord, virus, virus_coord, splice_type, seqtype)


```



```{r}
# restrict to splice type

spliced_insertions = insertions_data %>% 
  filter(split > 0) %>% # ensure we have at least one split read to support breakpoint definition.
  filter(exon_insert_type == "at_ref_exon_boundary") %>%
  filter(splice_type != '.') 


spliced_insertions %>%
  select(sample_id, virus, contig, splice_type, Gene_ID)

# 579 spliced insertions

```

```{r}

write.table(spliced_insertions, file='spliced_insertions.tsv', quote=F, sep="\t", row.names=F)


```

```{r}
# count samples and genes where insertions are spliced

spliced_insertions %>%
  select(sample_id, virus, Gene_ID) %>% unique() %>% group_by(Gene_ID) %>% tally(name='num_samples_spliced_at_gene') %>%
  arrange(desc(num_samples_spliced_at_gene))

# 301 genes involved with spliced insertions
```


```{r}
# how many samples with spliced insertions?

spliced_insertions %>%
  select(sample_id) %>% unique() %>% nrow()

# 229 samples

```

```{r}

# what cohorts, projects, and viruses?
spliced_insertions %>%
  select(sample_id, virus, cohort, project) %>% unique() %>% 
  group_by(cohort, project, virus) %>% tally() %>% arrange(desc(n)) %>% ungroup() %>% mutate(n_frac=prop.table(n))



```


```{r}

# what cohorts, projects, and viruses?  Breakdown by fractions of samples
spliced_insertions %>%
  mutate(virus = ifelse(grepl("HPV", virus), "HPV", virus)) %>% 
  select(sample_id, virus, cohort, project) %>% unique() %>% 
  group_by(virus) %>% tally() %>% arrange(desc(n)) %>% ungroup() %>% mutate(n_frac=prop.table(n))



```


```{r}
## What are the EBV spliced insertions?

spliced_insertions %>% filter(virus == "Eppstein-Barr") %>% select(sample_id, project, cohort, Gene_ID) %>% unique() %>%
  group_by(project, cohort, Gene_ID) %>% tally() %>% arrange(desc(n))


```


```{r}

# HPV spliced insertions?

spliced_insertions %>% filter(grepl("HPV", virus) ) %>% select(sample_id, project, cohort, Gene_ID) %>% unique() %>%
  group_by(project, cohort, Gene_ID) %>% tally() %>% arrange(desc(n))

```


```{r}

# murine retroviruses

spliced_insertions %>% filter(grepl("XMRV|urine", virus) ) %>% select(sample_id, project, cohort, Gene_ID, virus) %>% unique() %>%
  group_by(project, cohort, virus, Gene_ID) %>% tally() %>% arrange(desc(n))

```


```{r}
# HBV spliced insertions?

spliced_insertions %>% filter(grepl("HBV", virus) ) %>% select(sample_id, project, cohort, Gene_ID) %>% unique() %>%
  group_by(project, cohort, Gene_ID) %>% tally() %>% arrange(desc(n))

```



```{r}
# interesting splice into TP53:

spliced_insertions %>% filter(Gene_ID == "TP53^ENSG00000141510.14")


# looks like first exon of TP53 splices into the gag-pol into a cryptic splice on the revcomp of the 3' end of the gag/pol orf on the virus.

```

# Examine virus spliced sites involved


```{r}
# HPV16

spliced_insertions %>% filter(virus=="HPV16") %>% select(sample_id, virus_coord, human_coord) %>% unique() %>% 
  group_by(virus_coord) %>%  tally() %>% arrange(desc(n)) %>% ungroup() %>% mutate(n_frac=prop.table(n))


```

Notes from Jack:

227 =  E6*I 5’ss; also occasionally spliced to human 3’ss sequences downstream of integrated HPV16 DNA in lesions. 

408 =  E6*I 3’ss

578 =  This one is very puzzling.  The sequence does not look like a 5’ss (GT) or 3’ss (YAG) around this position (CACCTACA, 578 underlined).  Is it possible that there are variants of HPV16 with a C to G substitution to CAGCTACA that could create a 3’ss (CAG) for a E6* variant?  The problem with that though is the G at the 3’ end of the intron would be nucleotide 577 not 578, so I doubt that is the explanation.  Such a mutation could be checked two ways:  1) the RNAseq reads for the E6*I isoforms should cover position 577 downstream of 409, or 2) if the DNA sequences of the HPV16 from the tumors that had those 10 junctions were determined by hybridization capture and available.  This one is weird to me.

741 =  I think this is the E6*II 3’ss. 

881 =  E1^E4 5’ss;  also often spliced to human 3’ss sequences downstream of integrated HPV16 DNA in lesions.

3357 =  E4 3’ss.   881 and 3357 are the first and last nucleotides of the standard E1^E4 intron that splices E1 and E4 in frame.  (TAG)

3360 =  alternative E4 3’ss, 3 nucleotides downstream of the main one at 3357.  (CAG)

2708 =  Likely E2 3’ss. (AAG though).  It is 48 nucleotides upstream of the E2 ORF ATG.  The cognate 5’ss should be 881.  Note that there are several more AAG’s and one CAG between 2708 and  the start codon for E2 at 2756.  I am not sure what is known about 3’ss usage for E2, i.e. - Is 2708 the usual site?  Or one of the others?  Or maybe all or a set of them??

3633 =  This looks like a bona fide 5’ss. (GT). It is located in the E2 ORF, 13 nucleotides downstream of the stop codon of the E2 overlapping E4 ORF.  I am not familiar with this ss for standard HPV splicing.  It would be very interesting to see what 3’ss (3’ss’s?) these 6 junctions are joined to.  My guess is that they are in human sequences downstream of integrated HPV16 DNAs or they are joined to 5638.

5638 =  This looks like a bona fide 3’ss. (CAG).  Beyond that, it is puzzling.  It is located in the L2 ORF just 20 nucleotides upstream of the L2 stop codon.  Again, I am not familiar with this ss for standard HPV splicing.  The L1 ORF starts at 5639.  My guess is that this 3’ss is used for for L1 translation during viral replication by splicing from an upstream 5’ss in the viral genome?  If so, my guess is 3633 is the 5’ss at least part of the time.  Maybe the tumors with the 5638 3’ss junction had active HPV infections ongoing??  RNAseq might show full coverage of the viral genome if so.  If not, then I have two wild guesses what this 3’ss did.  One is that human sequences might occasionally splice from a human 5’ss into this 3’ss upstream of the UTR in integrated HPV16 DNA; and if so it then the 5’ ss’s are for these 9 junctions should be upstream in HPV16.  The other is that these might be very unusual HPV16 DNA integrations where the entire stretch of E6-E7-E1-E2-E5-and L2 (and maybe L1) are downstream of the URR; and if that is the case, then it would be interesting to see if the last 20 nucleotides of the L2 ORF and part or all of L1 are present in RNAseq data, or maybe there’s a polyA signal or a 5’ss further but unidentified 5’ss downstream in HPV16.  




```{r}
# HBV

spliced_insertions %>% filter(virus=="HBV") %>% select(sample_id, virus_brkend_grp, human_coord) %>% unique() %>% 
  group_by(virus_brkend_grp) %>%  tally() %>% arrange(desc(n)) %>% ungroup() %>% mutate(n_frac=prop.table(n))


```

```{r}
# EBV

spliced_insertions %>% filter(virus=="Eppstein-Barr") %>% select(sample_id, virus_brkend_grp,  human_coord) %>% unique() %>% 
  group_by(virus_brkend_grp) %>%  tally() %>% arrange(desc(n)) %>% ungroup() %>% mutate(n_frac=prop.table(n))


```





