---
title: "virus content and insertions analysis"
author: "Brian Haas"
date: "6/22/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(cowplot)

```


# sample info
```{r}

rnaseq_samples = read.table("data/SRA189004.acc_list.rnaseq", header=F, stringsAsFactors = F)[,1]
wgs_samples = read.table("data/SRA180295.acc_list.WGS", header=F, stringsAsFactors = F)[,1]
hivid_samples = read.table("data/SRA189003.acc_list.hivid", header=F, stringsAsFactors = F)[,1]

samples_info = bind_rows(data.frame(sample_type="rnaseq", sample_name=rnaseq_samples),
                         data.frame(sample_type="wgs", sample_name=wgs_samples),
                         data.frame(sample_type="hivid", sample_name=hivid_samples))



```


```{r}
sra_info = bind_rows(read.table("data/SraRunTable.hivid", header=T, sep=",", stringsAsFactors = F),
                     read.table("data/SraRunTable.rnaseq", header=T, sep=",", stringsAsFactors = F),
                     read.table("data/SraRunTable.WGS", header=T, sep=",", stringsAsFactors = F) ) %>%
    select(Run, Sample.Name, Bases, AvgSpotLen) %>%
    rename(sample_name = Run, sample_label = Sample.Name, ) %>%
    mutate(sra_num_reads = Bases / AvgSpotLen) %>%
  select(-Bases, -AvgSpotLen)

samples_info = left_join(samples_info, sra_info, by='sample_name') 


samples_info$sample_label_short = sapply(samples_info$sample_label, function(x) { str_split(x, "-")[[1]][4] })

write.table(samples_info, file="NatGen2015.samples_info.tsv", quote=F, sep="\t", row.names=F)

```



```{r}

# get align stats

read_align_stats = read.table(gzfile("data/NatGen2015.star_align_stats.tsv.gz"), header=T, sep="\t", stringsAsFactors = F)

read_align_stats = read_align_stats %>% select(sample_name, Number.of.input.reads) %>% rename(sample_read_count = Number.of.input.reads)

read_align_stats

```


```{r}

samples_info_w_align_stats = full_join(samples_info, read_align_stats, by='sample_name')

# verify our downloaded read counts match expectations
samples_info_w_align_stats %>% filter(sra_num_reads != sample_read_count)

```

```{r}

natgen_S4 = read.table("data/NatGen_sample_info_TableS4.txt", header=T, sep="\t", stringsAsFactors = F)

```



## refined insertions data
```{r}


insertions_data = read.table(gzfile("data/NatGenet2015.virus_insertions.tsv.gz"), sep="\t", stringsAsFactors = F, header=T, com='')

insertions_data = insertions_data %>% rename(sample_name = X.sample)

insertions_data = insertions_data %>% filter(total > 0)

insertions_data = full_join(samples_info, insertions_data, by='sample_name')

```


```{r}

# exclude the insertions at low complexity sequences

insertions_data = insertions_data %>%  filter(flankA_fU >= 0.5 & flankB_fU >= 0.5)
  
  
```  


```{r}

# assign virus column and specific coordinate info for virus and human

insertions_data$virus = sapply(insertions_data$virus_brkend_grp, function(x) { str_split(x, ":")[[1]][1] })

insertions_data = insertions_data %>% 
  mutate(humanchr = ifelse(chrA==virus, chrB, chrA)) %>%
  mutate(virus_coord = ifelse(chrA==virus, coordA, coordB)) %>%
  mutate(human_coord = ifelse(chrA==virus, coordB, coordA)) 


```




```{r}

# add normalizations for sample_read_count

insertions_data = left_join(insertions_data, read_align_stats, by='sample_name')

insertions_data = insertions_data %>% mutate(total_rpm = total / sample_read_count * 1e6)


```



```{r}

# remove K113 - will be analyzed separately

K113_insertions_data = insertions_data %>% filter(grepl("K113", virus))

insertions_data = insertions_data %>% filter(! grepl("K113", virus))


```


```{r}

# write refined insertion report

write.table(insertions_data, file="insertions.tsv", quote=F, sep="\t", row.names=F)
write.table(K113_insertions_data, file="K113.insertions.tsv", quote=F, sep="\t", row.names=F)

```

## Virus content


```{r}
virus_content_data = read.table(gzfile("data/NatGenet2015.virus_content.tsv.gz"), header=T, sep="\t", stringsAsFactors = F, com='')

virus_content_data = virus_content_data %>% rename(sample_name = X.sample)

virus_content_data = full_join(samples_info, virus_content_data, by='sample_name')

```



```{r}
# normalize virus content by sample read count

virus_content_data = left_join(virus_content_data, read_align_stats, by='sample_name')

virus_content_data = virus_content_data %>% mutate(mapped_rpm = mapped / sample_read_count * 1e6)


```



```{r}

# remove K113 here too, will be analyzed separately.

K113_virus_content_data = virus_content_data %>% filter(grepl("K113", virus))

virus_content_data = virus_content_data %>% filter(! grepl("K113", virus))


write.table(virus_content_data, "virus_content.tsv", quote=F, sep="\t", row.names=F)
write.table(K113_virus_content_data, "K113_virus_content.tsv", quote=F, sep="\t", row.names=F)

```



# Determine background / filtering levels for virus insertions

```{r}

#plot number of samples vs. min number of virus chimeric reads threshold.
min_chim_read_thresholds = c(0, 1, 5, 10, 20, 50, 100)
sample_counts_vs_min_chim_read_thresh = data.frame()
for (min_chim_read_threshold in min_chim_read_thresholds) {
    num_samples = insertions_data %>% filter(total >= min_chim_read_threshold) %>% 
        select(sample_name) %>% unique() %>% nrow()
    
    sample_counts_vs_min_chim_read_thresh = bind_rows(sample_counts_vs_min_chim_read_thresh, data.frame(min_chim_read_threshold=min_chim_read_threshold, num_samples=num_samples))
}


sample_counts_vs_min_chim_read_thresh %>% ggplot(aes(x=min_chim_read_threshold, y=num_samples)) + geom_point() + geom_line() + ylim(0, NA) + geom_vline(xintercept = 10, color='blue') + ggtitle("Sample count vs. min virus chimeric reads")


```


```{r}

#plot number of samples vs. min number of virus chimeric reads threshold.
min_chim_rpm_thresholds = c(0, 0.1, 0.2, 0.5, 1, 2)
sample_counts_vs_min_chim_rpm_thresh_by_sample = data.frame()
for (min_chim_rpm_threshold in min_chim_rpm_thresholds) {
    sample_count_by_sample_type = insertions_data %>% filter(total_rpm >= min_chim_rpm_threshold) %>% 
        select(sample_name, sample_type) %>% unique() %>% group_by(sample_type) %>% tally(name='num_samples') %>% mutate(min_chim_rpm_threshold=min_chim_rpm_threshold)
    
    sample_counts_vs_min_chim_rpm_thresh_by_sample = bind_rows(sample_counts_vs_min_chim_rpm_thresh_by_sample, sample_count_by_sample_type)
}


sample_counts_vs_min_chim_rpm_thresh_by_sample %>% ggplot(aes(x=min_chim_rpm_threshold, y=num_samples, color=sample_type)) + geom_point() + geom_line()  +
  geom_vline(xintercept = 0.2, color='blue') + 
  ggtitle("Sample count vs. min virus chimeric rpm") +
  facet_wrap(~sample_type, scale='free_y')




```


# Determine background / filtering levels for virus content

```{r}

virus_content_data %>% ggplot(aes(x=mapped)) + geom_density() + xlim(0,100)


```


```{r}

virus_mapped_read_counts = virus_content_data %>% select(sample_name, virus, mapped, n_bases_covered) %>%
    filter(! grepl("HPV", virus))

quantile(virus_mapped_read_counts$mapped, probs = seq(0,1,0.1), na.rm=T)
quantile(virus_mapped_read_counts$n_bases_covered, probs = seq(0,1,0.1), na.rm=T)

virus_mapped_read_counts
```

```{r}

#plot number of samples vs. min number of virus mapped reads threshold.
min_read_thresholds = c(0, 1, 5, 10, 20, 50, 100, 200, 500)
sample_counts_vs_min_read_thresh = data.frame()
for (min_read_threshold in min_read_thresholds) {
    num_samples = virus_content_data %>% filter(mapped >= min_read_threshold) %>% select(sample_name) %>% unique() %>% nrow()
    sample_counts_vs_min_read_thresh = bind_rows(sample_counts_vs_min_read_thresh, data.frame(min_read_threshold=min_read_threshold, num_samples=num_samples))
}


sample_counts_vs_min_read_thresh %>% ggplot(aes(x=min_read_threshold, y=num_samples)) + geom_point() + geom_line() +
    geom_vline(xintercept = 50, color='blue') + ggtitle("Sample count vs. min virus-mapped reads")

```
```{r}

#plot number of samples vs. min number of virus mapped reads RPM threshold.
min_mapped_rpm_thresholds =  c(1, 2, 3, 4, 5, 10, 20, 50, 100)
sample_counts_vs_min_mapped_rpm_thresh_by_sample = data.frame()
for (min_mapped_rpm_threshold in min_mapped_rpm_thresholds) {
    sample_count_by_sample_type = virus_content_data %>% filter(mapped_rpm >= min_mapped_rpm_threshold) %>% 
        select(sample_name, sample_type) %>% unique() %>% group_by(sample_type) %>% tally(name='num_samples') %>% mutate(min_mapped_rpm_threshold=min_mapped_rpm_threshold)
    
    sample_counts_vs_min_mapped_rpm_thresh_by_sample = bind_rows(sample_counts_vs_min_mapped_rpm_thresh_by_sample, sample_count_by_sample_type)
}


sample_counts_vs_min_mapped_rpm_thresh_by_sample %>% ggplot(aes(x=min_mapped_rpm_threshold, y=num_samples, color=sample_type)) + geom_point() + geom_line()  +
  geom_vline(xintercept = 20, color='blue') + 
  ggtitle("Sample count vs. min virus mapped rpm") +
  facet_wrap(~sample_type, scale='free_y')




```

```{r}

#plot number of samples vs. min number of virus mapped RPM threshold.
min_rpm_thresholds = c(1, 2, 3, 4, 5, 10, 20, 50, 100, 200, 500)
sample_counts_vs_min_rpm_thresh = data.frame()
for (min_rpm_threshold in min_rpm_thresholds) {
    num_samples = virus_content_data %>% filter(mapped_rpm >= min_rpm_threshold) %>% select(sample_name) %>% unique() %>% nrow()
    sample_counts_vs_min_rpm_thresh = bind_rows(sample_counts_vs_min_rpm_thresh, data.frame(min_rpm_threshold=min_rpm_threshold, num_samples=num_samples))
}


sample_counts_vs_min_rpm_thresh %>% ggplot(aes(x=min_rpm_threshold, y=num_samples)) + geom_point() + geom_line() +
    geom_vline(xintercept = 20, color='blue') + ggtitle("Sample count vs. min virus-mapped rpm")

```




```{r}

#plot number of samples vs. min n_bases_covered threshold.
min_covered_thresholds = c(0, 1, 5, 10, 20, 50, 100, 200, 300, 400, 500, 1000, 2000, 3000)
sample_counts_vs_min_covered_thresh = data.frame()
for (min_covered_threshold in min_covered_thresholds) {
    num_samples = virus_content_data %>% filter(n_bases_covered >= min_covered_threshold) %>% select(sample_name) %>% unique() %>% nrow()
    sample_counts_vs_min_covered_thresh = bind_rows(sample_counts_vs_min_covered_thresh, data.frame(min_covered_threshold=min_covered_threshold, num_samples=num_samples))
}


sample_counts_vs_min_covered_thresh %>% ggplot(aes(x=min_covered_threshold, y=num_samples)) + geom_point() + geom_line() +
    geom_vline(xintercept = 500, color='blue') + ggtitle("Sample count vs. min virus-covered bases")

```


```{r}

#plot number of samples vs. min n_bases_covered threshold.
min_covered_thresholds = c(0, 1, 5, 10, 20, 50, 100, 200, 300, 400, 500, 1000, 2000, 3000)
sample_counts_vs_min_covered_thresh = data.frame()
for (min_covered_threshold in min_covered_thresholds) {
    num_samples = virus_content_data %>% filter(n_bases_covered >= min_covered_threshold) %>% select(sample_name, sample_type) %>% unique() %>% group_by(sample_type) %>% tally(name='num_samples') %>% mutate(min_covered_threshold = min_covered_threshold)
    
    sample_counts_vs_min_covered_thresh = bind_rows(sample_counts_vs_min_covered_thresh, num_samples) 

}


sample_counts_vs_min_covered_thresh %>% ggplot(aes(x=min_covered_threshold, y=num_samples)) + geom_point() + geom_line() +
    geom_vline(xintercept = 500, color='blue') + ggtitle("Sample count vs. min virus-covered bases") + facet_wrap(~sample_type, scale='free')

```





# Filter insertions and virus content accoring to decided-upon thresholds.

```{r}

MIN_CHIM_READS = 10
MIN_MAPPED_READS = 50
MIN_N_BASES_COVERED = 500

```


## filter insertions


```{r}

filtered_insertions_data = insertions_data %>% 
    filter(total >= MIN_CHIM_READS) %>%
    rowwise() %>% mutate(i_site = paste(virus, humanchr, sep=":"))

write.table(filtered_insertions_data, file="filtered_insertions.tsv", quote=F, sep="\t", row.names=F)

filtered_insertions_data

```


```{r}

# how many primary insertions and with reasonable kmer complexity at junctions?

filtered_insertions_data %>% filter(is_primary == "True") 

filtered_insertions_data %>% filter(is_primary == "True") %>% filter(flankA_fU >= 0.3 & flankB_fU >= 0.3)

filtered_insertions_data %>% filter(is_primary == "True") %>% filter(flankA_fU >= 0.3 & flankB_fU >= 0.3) %>% select(sample_name, virus) %>% unique() %>% arrange(sample_name) 

```




## filter virus content


```{r}

filtered_virus_content_data = virus_content_data %>% 
  filter(mapped >= MIN_MAPPED_READS & n_bases_covered >= MIN_N_BASES_COVERED) 

write.table(filtered_virus_content_data, "filtered_virus_content.tsv", sep="\t", row.names = F, quote=F)

head(filtered_virus_content_data)
```





```{r}

# include filtering criteria


p2 = filtered_virus_content_data %>% group_by(virus) %>% tally() %>% ggplot(aes(x=virus, y=n)) + geom_col() +
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) + 
  geom_text(aes(label = n), vjust = 'center', color='orange')

#plot(p2)

p = filtered_virus_content_data %>% ggplot(aes(x=virus, y=mapped)) + geom_boxplot() + 
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5, size=rel(0.7) ))
#plot(p)



p = plot_grid(p2, p, ncol=1, align='v', rel_heights = c(0.25, 0.75))
plot(p)


```


# integrate virus content and insertion data


```{r}

#Examine numbers of samples per tissue type


tissue_sample_counts = samples_info %>% group_by(sample_type) %>% tally(name='tot_samples')

tissue_sample_counts


```

```{r}
# plot sample counts

tissue_sample_counts %>%
    ggplot(aes(x=sample_type, y=tot_samples)) + geom_bar(position='dodge', stat='identity') +
    geom_text(aes(label = tot_samples )  , vjust = 'left', color='black') +
    theme(axis.text.x = element_text(angle = 90, hjust = 1))
               



```










```{r}

filtered_insertions_distilled = filtered_insertions_data %>% 
    # take highest scoring i_site 
    group_by(sample_name, i_site) %>% arrange(desc(total)) %>% filter(row_number() == 1) %>% ungroup() %>%
    # add the total to the i_site
    mutate(i_site = paste(i_site, paste0("T=",total), sep="^")) %>%
    
    # summarize i_sites per sample
    group_by(sample_name) %>%
    mutate(i_sites = paste(i_site, collapse=",")) %>% 
    select(sample_type, sample_name, i_sites) %>% 
    unique()
    

```

```{r}

filtered_insertions_distilled

```




```{r}

filtered_virus_content_distilled = filtered_virus_content_data %>% 
    mutate(virus_token = paste0(virus, "^Mpd=", mapped, "^Bcvd=", n_bases_covered, "^Fr=", frac_covered)) %>%
    group_by(sample_name) %>% 
    arrange(desc(frac_covered)) %>%
    mutate(virus_list = paste(virus_token, collapse=',') ) %>% 
    ungroup() %>% 
    select(sample_name, sample_type, virus_list) %>% unique()

filtered_virus_content_distilled

```


```{r}

merged_distilled_report = full_join(filtered_virus_content_distilled, filtered_insertions_distilled, by=c('sample_name', 'sample_type'))

merged_distilled_report


```

```{r}
# QC: are there insertions present with no viruses detected? Ideally very few.
merged_distilled_report %>% filter(is.na(virus_list) & ! is.na(i_sites))
```

```{r}

write.table(merged_distilled_report, file='NatGen2015.filtered_virus_content_and_insertions.distilled.tsv', quote=F, sep="\t", row.names=F)


```



```{r}

#More QC: are we finding discrepancies between viruses and insertions?

row_processor = function(virus_list, isite_list) {
    
    virus_list = str_split(virus_list, ",")[[1]]
    #print(virus_list)
    viruses = unique(sapply(virus_list, function(x) { str_split(x, "\\^")[[1]][1]}))
    
    #print(viruses)
    
    isite_list = str_split(isite_list, ",")[[1]]
    isite_viruses = unique(sapply(isite_list, function(x) { str_split(x, ":")[[1]][1]}))
    
    #print(isite_viruses)
    
    isite_virus_not_contained = isite_viruses[ ! isite_viruses %in% viruses]
    #isite_virus_not_contained = c(isite_viruses, viruses)
    
    #print(isite_virus_not_contained)
    
    retval = paste(isite_virus_not_contained, collapse=",")
    #print(retval)
    if (retval %in% c("", "NA") ) {
        retval = NA
    }
    
    return(retval)
    
}

```

```{r}

merged_report_w_conflicts = merged_distilled_report %>% rowwise() %>% mutate(conflict = row_processor(virus_list, i_sites)) %>% ungroup()

merged_report_w_conflicts 

```


```{r}

# pull out the conflicts

merged_report_w_conflicts %>% filter(! is.na(conflict) )


```


```{r}

# plot all insertions ranked desc(total), indicate positions of discrepancies

# first restructure the reporting for analysis

combined_data = full_join(filtered_virus_content_data, filtered_insertions_data, by=c('sample_name', 'virus', 'sample_type'), suffix=c('_vcontent', '_insertion') )


```

```{r}
combined_with_insertions_data = combined_data %>% 
    filter(! is.na(total)) %>%
    mutate(orphan = is.na(mapped))


p = combined_with_insertions_data %>%
    mutate(row=row_number()) %>% 
    ggplot(aes(y=total, x=reorder(row, -total))) + geom_point(aes(color=orphan)) +
    scale_y_continuous(trans='log10')

plot(p)


```

```{r}

combined_with_insertions_data %>% filter(orphan) %>% select(sample_name, virus, i_site, total)


```


```{r}

# count total number of samples w/ virus

sample_counts_with_virus = combined_data %>% filter(! is.na(mapped)) %>% 
    select(sample_type, sample_name) %>% unique() %>% 
    group_by(sample_type) %>% tally(name='w_virus')


sample_counts_with_virus

```



```{r}

sample_counts_with_insertions =  combined_data %>% filter(! is.na(total)) %>% 
    select(sample_type, sample_name) %>% unique() %>% group_by(sample_type) %>% tally(name='w_inserts')


sample_counts_with_insertions
```

```{r}

# which samples have the candidate insertions?
combined_data %>% filter(! is.na(total)) %>% 
    select(sample_type, sample_name, virus, total) %>% group_by(sample_type, sample_name, virus) %>% arrange(desc(total)) %>%
  filter(row_number() == 1) %>% ungroup()


```





```{r}

summary_count_table = full_join(tissue_sample_counts, sample_counts_with_virus)
summary_count_table = full_join(summary_count_table, sample_counts_with_insertions)

summary_count_table[is.na(summary_count_table)] = 0

summary_count_table
```




```{r}

summary_count_table_w_frac = summary_count_table %>% 
    mutate(frac_w_virus = w_virus/tot_samples,
           frac_w_inserts = w_inserts / tot_samples)

summary_count_table_w_frac
```



```{r}


natgen_virus_content_summary_plot = summary_count_table_w_frac %>% gather(key=frac_type, value=fraction, frac_w_virus, frac_w_inserts) %>%
    ggplot(aes(x=frac_type, y=fraction, fill=frac_type)) + geom_bar(position='dodge', stat='identity') +
    geom_text(aes(label =  sprintf("%.2f", fraction) )  , vjust = 'left', color='black') +
    theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
    facet_wrap(~sample_type, scale='free_x')



natgen_virus_content_summary_plot
```

```{r}

ggsave(natgen_virus_content_summary_plot, file="natgen_virus_summary_plot.pdf", width=12, height=12)



```



```{r}

# organize into a matrix of median reads mapped by tissue_id

virus_median_mapped_by_tissue_id = filtered_virus_content_data %>% select(sample_type, virus, mapped) %>% group_by(sample_type, virus) %>% 
  mutate(median_mapped = median(mapped), sample_count = n()) %>% 
  select(sample_type, virus, median_mapped, sample_count) %>% unique() 



```

```{r}

virus_median_mapped_all_plot = virus_median_mapped_by_tissue_id  %>% ggplot(aes(y=sample_type, x=median_mapped, size=sample_count, color=virus)) +
  geom_point() +
  scale_x_log10()

virus_median_mapped_all_plot
```

```{r}
ggsave(virus_median_mapped_all_plot, file="natgen_virus_median_mapped_all_plot.pdf", width=16, height=12)

```






